<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
    <meta charset="utf-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><link rel="canonical" href="https://hpc2n.github.io/selecting-modules/programs/" />
      <link rel="shortcut icon" href="../img/favicon.ico" />
    <title>Software module examples - Selecting software modules</title>
    <link rel="stylesheet" href="../css/theme.css" />
    <link rel="stylesheet" href="../css/theme_extra.css" />
        <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/styles/github.min.css" />
        <link href="../assets/_markdown_exec_pyodide.css" rel="stylesheet" />
        <link href="../assets/_markdown_exec_ansi.css" rel="stylesheet" />
        <link href="../assets/_mkdocstrings.css" rel="stylesheet" />
        <link href="../css/extra.css" rel="stylesheet" />
    
      <script>
        // Current page data
        var mkdocs_page_name = "Software module examples";
        var mkdocs_page_input_path = "programs.md";
        var mkdocs_page_url = "/selecting-modules/programs/";
      </script>
    
    <!--[if lt IE 9]>
      <script src="../js/html5shiv.min.js"></script>
    <![endif]-->
      <script src="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/11.8.0/highlight.min.js"></script>
      <script>hljs.highlightAll();</script> 
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
    <div class="wy-side-scroll">
      <div class="wy-side-nav-search">
          <a href="..">
          <img src="../images/logo-modules-course_v2.png" class="logo" alt="Logo"/>
        </a><div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
      <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="..">Home</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../intro/">The module system</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../commands/">Module system commands</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../toolchains/">Compiler toolchains</a>
                </li>
              </ul>
              <ul class="current">
                <li class="toctree-l1 current"><a class="reference internal current" href="#">Software module examples</a>
    <ul class="current">
    <li class="toctree-l2"><a class="reference internal" href="#outline">Outline</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#packages">Packages</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#python-based__packages">Python-based packages</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#bundles">Bundles</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#principles">Principles</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#python__ides">Python IDEs</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#r">R</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#principles_1">Principles</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#r-based__packages">R-based packages</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#principles_2">Principles</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#rstudio">RStudio</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#popular__tools">Popular tools</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#principle">Principle</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#specialised__applications">Specialised Applications</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#exercises__stop__recording">Exercises (Stop recording)</a>
        <ul>
    <li class="toctree-l3"><a class="reference internal" href="#python">Python</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#r_1">R</a>
    </li>
    <li class="toctree-l3"><a class="reference internal" href="#matlab__and__another__tool__you__will__be__using">Matlab and another tool you will be using</a>
    </li>
        </ul>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#wrap-up__with__questions__no__recording">Wrap-up with questions (No recording)</a>
    </li>
    <li class="toctree-l2"><a class="reference internal" href="#courses">Courses</a>
    </li>
    </ul>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../batch/">Modules in batch scripts</a>
                </li>
              </ul>
              <ul>
                <li class="toctree-l1"><a class="reference internal" href="../summary/">Summary</a>
                </li>
              </ul>
              <p class="caption"><span class="caption-text">Extra</span></p>
              <ul>
                  <li class="toctree-l1"><a class="reference internal" href="../advanced_module/">Advanced module commands</a>
                  </li>
              </ul>
      </div>
    </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">
      <nav class="wy-nav-top" role="navigation" aria-label="Mobile navigation menu">
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="..">Selecting software modules</a>
        
      </nav>
      <div class="wy-nav-content">
        <div class="rst-content"><div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href=".." class="icon icon-home" aria-label="Docs"></a></li>
      <li class="breadcrumb-item active">Software module examples</li>
    <li class="wy-breadcrumbs-aside">
    </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
            <div class="section" itemprop="articleBody">
              
                <h1 id="examples__of__specific__tools">Examples of specific tools<a class="headerlink" href="#examples__of__specific__tools" title="Permanent link">&para;</a></h1>
<div class="admonition note">
<p class="admonition-title">Objectives</p>
<ul>
<li>Be able to find some typical software modules that are installed </li>
<li>Learn how to load those software modules (and any prerequisites)</li>
<li>Be able to find out if <strong>Python packages</strong> are installed as their own modules</li>
<li>Learn how to find bundles of <strong>R packages</strong> installed as modules </li>
</ul>
</div>
<p>In the previous section we looked at toolchains, which are often used to install or build your own software. In this chapter we will look at some examples of how to find and load some of the typical software modules that are installed at many of the Swedish HPC centres. </p>
<p>Loading modules works the same whether the modules are toolchains or standalone packages, with and without prerequisites. The procedure is usually some variation on the following:</p>
<ol>
<li>Use <code>ml spider &lt;package&gt;</code> to see <strong>whether a module is installed</strong>, and if so, view the available versions. (At facilities that do not hide packages depending on prerequisites, <code>module avail &lt;package&gt;</code> may be preferred.)</li>
<li>Use <code>ml spider &lt;package&gt;/&lt;version&gt;</code> to <strong>view prerequisites</strong> for a specific version of the module, if any.</li>
<li><strong>Load</strong> the modules with <code>ml &lt;prerequisite&gt;/&lt;version&gt; &lt;package&gt;/&lt;version&gt;</code>.</li>
</ol>
<p>Most of the examples below use outputs from Cosmos, but the workflows will be similar at other institutions except where otherwise noted.</p>
<h3 id="outline">Outline<a class="headerlink" href="#outline" title="Permanent link">&para;</a></h3>
<ul>
<li>Principles of making Python and R <strong>packages</strong> available to scripts and interactive analysis.</li>
<li>Start <strong>popular tools</strong> like MATLAB and OpenFOAM and notes about licensed software</li>
<li><strong>Exercise</strong>: Dig deeper into <strong>your favorite tools</strong></li>
</ul>
<h2 id="packages">Packages<a class="headerlink" href="#packages" title="Permanent link">&para;</a></h2>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Both R and Python packages often include many so-called extensions (dependent packages that are usually called modules when not dealing with LMOD modules) that are installed, but cannot be found with <code>ml spider</code> or <code>ml avail</code>.</p>
<p>In these cases, if you have loaded at least the prerequisites of a standalone package containing the extension, you can use <code>ml show &lt;package&gt;</code> on the standalone package to view the Lua module file, which usually has a section on included extensions near the top.</p>
<p>For example, NumPy, SciPy, and Pandas, among other packages, are all are included in <code>SciPy-bundle</code>. If you have at least loaded a GCC version, you can use <code>ml show SciPy-bundle</code> to view all of the included extensions (Python modules) in the compatible SciPy-bundle.</p>
</div>
<h2 id="python-based__packages">Python-based packages<a class="headerlink" href="#python-based__packages" title="Permanent link">&para;</a></h2>
<p>We already learned how to load Python module of a specific version.</p>
<p>It varies between clusters how many packages are installed with the base Python packages, how many are installed as separate modules, what prerequisites are required (most centres apart from PDC are the same), and the available version numbers. </p>
<details class="note">
<summary>Packages at the clusters</summary>
<ul>
<li><strong>HPC2N</strong> and <strong>C3SE:</strong> Little is installed with the Python module, but most of the common Python packages are available as extra modules (SciPy-bundle, Jupyter, mpi4py, matplotlib, tensorflow, PyTorch, Python-bundle-PyPi, &hellip;)</li>
<li><strong>LUNARC:</strong> Little is installed with the Python module, but instead most of the common Python packages are available as extra modules (SciPy-bundle, Jupyter, mpi4py, matplotlib, tensorflow, PyTorch, Python-bundle-PyPi, &hellip;). Anaconda3 bundles more into one module (SciPy, Pandas, Matplotlib, etc), but other modules loaded alongside it are typically not recognised. Additional modules must be installed in a custom environment.</li>
<li><strong>NSC:</strong> Little is installed with the Python module, but most of the common Python packages are available as extra modules (SciPy-bundle, matplotlib, mpi4py, PyTorch, Python-bundle-PyPi, Jupyter, &hellip;). Most programs on Tetralith also have an extra prerequisite, <code>buildtool-easybuild/4.X.X-hpc&lt;version&gt;</code> that must be loaded before anything else.  </li>
<li><strong>UPPMAX-Pelle:</strong> Little is installed with the Python module, but most of the common Python packages are available as extra modules (SciPy-bundle, matplotlib, mpi4py, Python-bundle-PyPi, Jupyter-lab, &hellip;). No Prerequisites are needed! </li>
<li><strong>UPPMAX-Bianca:</strong> Very many packages are installed with the Python module.  </li>
<li><strong>PDC:</strong> most modules included in SciPy-bundle (NumPy, SciPy, Pandas, etc.) are in <code>cray-python</code> modules, and these are compatible with a couple of the installed versions of matplotlib. Python modules that do not include the <code>cray-</code> prefix have very little installed in them and are not compatible with most other Python-adjacent modules; these are typically intended as bases for users to build their own environments. Most programs on Dardel also have an extra prerequisite, <code>PDC/XX.XX</code> or <code>PDCOLD/XX.XX</code> that must be loaded before anything else.</li>
</ul>
</details>
<h3 id="bundles">Bundles<a class="headerlink" href="#bundles" title="Permanent link">&para;</a></h3>
<ul>
<li>
<p>The bundle names reflect the content, like Python packages, and its version, but also which Python version, compilers and libraries that are compatible with it.</p>
</li>
<li>
<p>The module endings may contain GCCcore-X.Y.Z and/or [YEAR-a/b]. Example <code>SciPy-bundle/2024.05-gfbf-2024a</code> or <code>Python/3.12.3-GCCcore-13.3.0</code></p>
<ul>
<li>GCCcore reflects the GCC compiler version that is compatible when using C/C++ &ldquo;back end&rdquo; code.</li>
<li>The year reflects an EasyBuild toolchain, see <a href="https://docs.easybuild.io/common-toolchains/#common_toolchains_overview_foss">FOSS toolchains</a>.</li>
</ul>
</li>
</ul>
<details class="note">
<summary>Some FOSS tool chains and Python version using them</summary>
<table>
<thead>
<tr>
<th>FOSS</th>
<th>Python version</th>
<th>GCC version</th>
<th>Bundle version</th>
</tr>
</thead>
<tbody>
<tr>
<td>2024a</td>
<td>3.12.3</td>
<td>13.3.0</td>
<td>2024.06/06</td>
</tr>
<tr>
<td>2025b</td>
<td>3.13.5</td>
<td>14.3.0</td>
<td>2025.07</td>
</tr>
</tbody>
</table>
<ul>
<li><code>foss</code> is the full level toolchain.</li>
<li><code>gfbf</code> means that the libraries FlexiBLAS (incl. LAPACK) + FFTW are included.</li>
<li>
<p><code>gompi</code> means that the MPI library OpenMPI is included.</p>
</li>
<li>
<p>See <a href="https://docs.easybuild.io/common-toolchains/#toolchains_diagram">Toolchain diagram</a></p>
</li>
</ul>
</details>
<div class="admonition warning">
<p class="admonition-title">Warning</p>
<ul>
<li>Make sure to use bundles that are compatible with each-other and with needed Python version.</li>
<li>Otherwise it is better to create isolated environments with Conda or virtual environments, see <a href="python_virtual_environments.md">Virtual environments in Python</a>.</li>
</ul>
</div>
<details class="note">
<summary>Some well-known bundles</summary>
<ul>
<li>HPC and big data<ul>
<li>dask</li>
<li>mpi4py</li>
<li>numba</li>
</ul>
</li>
<li>Scientific tools<ul>
<li>SciPy-bundles: <code>numpy</code>, <code>pandas</code>, <code>scipy</code></li>
<li>xarray</li>
</ul>
</li>
<li>Biopython</li>
<li>Interactivity<ul>
<li>iPython</li>
<li>JupyterLab</li>
</ul>
</li>
<li>Graphics and diagrams<ul>
<li>Matplotlib</li>
<li>Seaborn</li>
</ul>
</li>
<li>Machine Learning<ul>
<li>scikit-learn</li>
<li>PyTorch</li>
<li>TensorFlow</li>
</ul>
</li>
<li>Bundle of useful packages<ul>
<li>Python-bundle-PyPI</li>
</ul>
</li>
</ul>
</details>
<p><a href="https://docs.uppmax.uu.se/software/python_bundles/">See also about Python bundles at Pelle</a></p>
<h3 id="principles">Principles<a class="headerlink" href="#principles" title="Permanent link">&para;</a></h3>
<ul>
<li>Decide what you need!<ol>
<li>Start new project with newest toolchain</li>
<li>Go for version you have used before (reproduce)</li>
<li>Exact versions of many packages may need an isolated environment.</li>
</ol>
</li>
<li>
<p>Load one or several bundles, python is loaded on the fly!</p>
</li>
<li>
<p>Check versions</p>
</li>
</ul>
<div class="language-console highlight"><pre><span></span><code><span id="__span-0-1"><a id="__codelineno-0-1" name="__codelineno-0-1" href="#__codelineno-0-1"></a><span class="go">ml spider matplotlib</span>
</span></code></pre></div>
<p>This is a very good way to find packages that are not in a bundle with the same name. For instance, <code>pandas</code> and <code>numpy</code> are parts of the <code>SciPy-bundle</code></p>
<p>or</p>
<div class="language-console highlight"><pre><span></span><code><span id="__span-1-1"><a id="__codelineno-1-1" name="__codelineno-1-1" href="#__codelineno-1-1"></a><span class="go">ml avail matplotlib</span>
</span></code></pre></div>
<ul>
<li>Load prerequisites, if needed, and then</li>
</ul>
<div class="language-console highlight"><pre><span></span><code><span id="__span-2-1"><a id="__codelineno-2-1" name="__codelineno-2-1" href="#__codelineno-2-1"></a><span class="go">ml matplotlib/&lt;version&gt;</span>
</span></code></pre></div>
<p>or</p>
<div class="language-console highlight"><pre><span></span><code><span id="__span-3-1"><a id="__codelineno-3-1" name="__codelineno-3-1" href="#__codelineno-3-1"></a><span class="go">ml matplotlib/&lt;version&gt;</span>
</span></code></pre></div>
<ul>
<li>Start Python session in a console with</li>
</ul>
<div class="language-console highlight"><pre><span></span><code><span id="__span-4-1"><a id="__codelineno-4-1" name="__codelineno-4-1" href="#__codelineno-4-1"></a><span class="go">python</span>
</span></code></pre></div>
<ul>
<li>Load a needed library, like</li>
</ul>
<div class="language-python highlight"><pre><span></span><code><span id="__span-5-1"><a id="__codelineno-5-1" name="__codelineno-5-1" href="#__codelineno-5-1"></a><span class="kn">import</span><span class="w"> </span><span class="nn">matplotlib</span>
</span></code></pre></div>
<div class="admonition note">
<p class="admonition-title">Isolated environments</p>
<ul>
<li>The course Python in an HPC environment: <a href="https://uppmax.github.io/HPC-python/day2/use_isolated_environments.html">isolated environments</a></li>
</ul>
</div>
<details class="warning">
<summary>Using Bianca?</summary>
<ul>
<li>There are no bundle modules on Bianca</li>
<li>But, the Python modules have very many packages installed already.</li>
</ul>
</details>
<div class="admonition tip">
<p class="admonition-title">Tip</p>
<p><code>grep</code> does not work directly on the outputs of module commands like <code>ml show &lt;module&gt;</code>. To search for an extension in a very long Lua module file, copy the full path of the <code>.lua</code> file from the <code>ml show</code> output, and use <code>less /path/to/module.lua | grep &lt;extension&gt;</code>. If there is no output, the extension is not present.</p>
</div>
<details class="note">
<summary>Missing a package?</summary>
<ul>
<li>Install by conda, pip or other tool or contact support</li>
<li><a href="https://uppmax.github.io/HPC-python/day2/install_packages.html">Install packages</a></li>
<li><a href="https://uppmax.github.io/HPC-python/day2/use_isolated_environments.html">Use isolated environments</a></li>
</ul>
</details>
<h3 id="python__ides">Python IDEs<a class="headerlink" href="#python__ides" title="Permanent link">&para;</a></h3>
<ul>
<li>If you are comfortable editing code in a basic text editor and running at the command-line, the modules used in the example above are all you need.</li>
<li>For more information on choosing and loading IDEs to work with Matplotlib graphics interactively, we refer readers to <a href="https://uppmax.github.io/HPC-python/day2/IDEs.html">this documentation from the Python for HPC course</a>.</li>
</ul>
<details class="note">
<summary>Python IDEs</summary>
<ul>
<li>Jupyter</li>
<li>Spyder</li>
<li>Visual Studio Code</li>
<li>HPC-python course<ul>
<li><a href="https://uppmax.github.io/HPC-python/day2/IDEs_cmd.html">Starting from command line</a></li>
<li><a href="https://uppmax.github.io/HPC-python/day2/ondemand-desktop.html">Starting from OnDemand</a></li>
</ul>
</li>
</ul>
</details>
<h2 id="r">R<a class="headerlink" href="#r" title="Permanent link">&para;</a></h2>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>Most facilities have only built a handful of R releases, so many of the dependent modules are adapted for multiple versions. The version of such any dependent package should always be specified to ensure reproducibility. If the version number is omitted, the latest will be loaded by default, and that version may change without warning.</p>
</div>
<p>Most HPC centres have prerequisites for R, but at a few, like Alvis, R can be loaded directly. Always check the prerequisites with <code>ml spider</code> or <code>ml avail</code>.</p>
<h3 id="principles_1">Principles<a class="headerlink" href="#principles_1" title="Permanent link">&para;</a></h3>
<ul>
<li>Check versions</li>
<li>Load prerequisites if needed and then R</li>
<li>Start R console with</li>
</ul>
<div class="language-console highlight"><pre><span></span><code><span id="__span-6-1"><a id="__codelineno-6-1" name="__codelineno-6-1" href="#__codelineno-6-1"></a><span class="go">R</span>
</span></code></pre></div>
<h3 id="r-based__packages">R-based packages<a class="headerlink" href="#r-based__packages" title="Permanent link">&para;</a></h3>
<p>At most HPC centres, the base R module usually contains relatively few extensions. Most of the popular packages are in additional bundles like the general <code>R-bundle-CRAN</code> and bundle for genomic tools <code>R-bundle-Bioconductor</code>. Most HPC centres have prerequisites for R, but at a few, like Alvis, R can be loaded directly. Always check the prerequisites with <code>ml spider</code> or <code>ml avail</code>.</p>
<details class="note">
<summary>Cluster-specific approaches</summary>
<ul>
<li><strong>NSC (Tetralith):</strong> R can be loaded directly, but <strong>contains few installed packages, and there are no bundles</strong> to provide more. Users are typically expected to install their own extension libraries. <strong>RStudio is included in the base R module, however</strong>.</li>
<li><strong>PDC (Dardel):</strong> Like most programs on Dardel, R also has the prerequisite <code>PDC/XX.XX</code> or <code>PDCOLD/XX.XX</code>, but the compiler and MPI library are chosen for you. There are about 250 packages available in the basic R module, and there are no additional bundles to provide more packages. Users are typically expected to install their own extension libraries. </li>
<li><strong>C3SE (Alvis):</strong> R can be loaded directly, but has few installed packages. RStudio can only be reached from an OnDemand session.</li>
<li><strong>UPPMAX (Bianca)</strong>: R can be loaded directly, but has few installed packages. More common modules are available as extensions with the <code>R_packages</code> module. RStudio is also a separate module.</li>
<li><strong>LUNARC:</strong> Little is installed with the basic R module, but most common packages are available as extensions of R-bundle-CRAN or R-bundle-Bioconductor. RStudio is also a separate module, and is available as an On-Demand application that automatically loads R and various bundles at start-up.</li>
<li><strong>UPPMAX (Pelle):</strong> Little is installed with the basic R module, but most common packages are available as extensions of R-bundle-CRAN or R-bundle-Bioconductor. RStudio is also a separate module.</li>
<li><strong>HPC2N:</strong> Little is installed with the basic R module, but most common packages are available as extensions of R-bundle-CRAN, R-bundle-CRAN-extra, or R-bundle-Bioconductor. RStudio is a separate module and only runs on the login nodes via Thinlinc, so it should be used sparingly.</li>
</ul>
</details>
<details class="note">
<summary>About packages on HPC systems</summary>
<p>Course module on <a href="https://uppmax.github.io/R-matlab-julia-HPC/r/packages/">R packages</a></p>
</details>
<p>Many R-packages conveniently specify the version of R they are compatible with in the module name. One example of this is Bioconductor.</p>
<h3 id="principles_2">Principles<a class="headerlink" href="#principles_2" title="Permanent link">&para;</a></h3>
<ul>
<li>Check versions</li>
</ul>
<div class="language-console highlight"><pre><span></span><code><span id="__span-7-1"><a id="__codelineno-7-1" name="__codelineno-7-1" href="#__codelineno-7-1"></a><span class="go">ml spider R-bundle</span>
</span></code></pre></div>
<p>or</p>
<div class="language-console highlight"><pre><span></span><code><span id="__span-8-1"><a id="__codelineno-8-1" name="__codelineno-8-1" href="#__codelineno-8-1"></a><span class="go">ml avail R-bundle</span>
</span></code></pre></div>
<ul>
<li>Load prerequisites, if needed, and then</li>
</ul>
<div class="language-console highlight"><pre><span></span><code><span id="__span-9-1"><a id="__codelineno-9-1" name="__codelineno-9-1" href="#__codelineno-9-1"></a><span class="go">ml R-bundle-CRAN/&lt;version&gt;</span>
</span></code></pre></div>
<p>or</p>
<div class="language-console highlight"><pre><span></span><code><span id="__span-10-1"><a id="__codelineno-10-1" name="__codelineno-10-1" href="#__codelineno-10-1"></a><span class="go">ml R-bundle-Bioconductor/&lt;version&gt;</span>
</span></code></pre></div>
<ul>
<li>The bundles should start the R interpreter ion the fly.</li>
<li>Start R session in a console with</li>
</ul>
<div class="language-console highlight"><pre><span></span><code><span id="__span-11-1"><a id="__codelineno-11-1" name="__codelineno-11-1" href="#__codelineno-11-1"></a><span class="go">R</span>
</span></code></pre></div>
<ul>
<li>Load a needed library, like</li>
</ul>
<div class="language-R highlight"><pre><span></span><code><span id="__span-12-1"><a id="__codelineno-12-1" name="__codelineno-12-1" href="#__codelineno-12-1"></a><span class="o">&gt;</span><span class="w"> </span><span class="nf">library</span><span class="p">(</span><span class="n">parallel</span><span class="p">)</span>
</span></code></pre></div>
<details class="note">
<summary>Missing a package?</summary>
<ul>
<li>Install by yourself or contact support</li>
<li><a href="https://uppmax.github.io/R-matlab-julia-HPC/r/packages/#installing-your-own-packages">Course material</a></li>
</ul>
</details>
<h3 id="rstudio">RStudio<a class="headerlink" href="#rstudio" title="Permanent link">&para;</a></h3>
<p>Many clusters have RStudio as a module but for Alvis and Dardel RStudio can only be reached from OnDemand.</p>
<details class="note">
<summary>RStudio from course</summary>
<p><a href="https://uppmax.github.io/R-matlab-julia-HPC/r/rstudio/">RStudio</a></p>
</details>
<h2 id="popular__tools">Popular tools<a class="headerlink" href="#popular__tools" title="Permanent link">&para;</a></h2>
<p>There are of course many popular tools among cluster users. Many of them can be found in every cluster, but some clusters are more niched and may lack some of these common tools.</p>
<p>Capitalisation and other naming conventions may vary between HPC centres.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>Are you lacking a tool? Ask the support to install it!</p>
</div>
<h3 id="principle">Principle<a class="headerlink" href="#principle" title="Permanent link">&para;</a></h3>
<ul>
<li>Check versions with <code>spider</code> or <code>avail</code></li>
<li>Load prerequisites, if needed, and then</li>
</ul>
<div class="language-console highlight"><pre><span></span><code><span id="__span-13-1"><a id="__codelineno-13-1" name="__codelineno-13-1" href="#__codelineno-13-1"></a><span class="go">ml &lt;module name&gt;</span>
</span></code></pre></div>
<ul>
<li>Start the tool by following documentation of the tool or the local cluster documentation.</li>
<li>Usually it is the tool name with lower case or with the first letter capitalised. Or, like for OpenFOAM, commands starting with <code>foam</code>.</li>
</ul>
<p>You may try out one or several tools in Exercise 3.</p>
<ul>
<li>There we also give a bit more detailed info about<ul>
<li>MATLAB</li>
<li>OpenFOAM </li>
</ul>
</li>
</ul>
<details class="note">
<summary>MATLAB in HPC course</summary>
<p><a href="https://uppmax.github.io/R-matlab-julia-HPC/matlab/intro-matlab/">Intro to MATLAB</a></p>
</details>
<p>??? note &ldquo;MATLAB Add-Ons and Toolboxes </p>
<div class="language-text highlight"><pre><span></span><code>These should be available through the Matlab GUI.
</code></pre></div>
<div class="admonition important">
<p class="admonition-title">Important</p>
<ul>
<li>The Matlab GUI and many other graphical tools is prone to hogging resources if not launched carefully, which makes it risky to run on a login node. </li>
<li>In general, the GUI should only be run via either Desktop On Demand or after booking interactive allocations on compute nodes with <code>salloc</code> or <code>interactive</code>. </li>
<li>For more particulars on <em>running</em> Matlab, see the <a href="https://uppmax.github.io/R-matlab-julia-HPC/matlab/load_runMatlab.html">relevant page of the R, Matlab, and Julia for HPC course materials</a>.</li>
</ul>
</div>
<h2 id="specialised__applications">Specialised Applications<a class="headerlink" href="#specialised__applications" title="Permanent link">&para;</a></h2>
<ul>
<li>For most specialised packages (Amber, GROMACS, Nextflow, VASP, etc),<ul>
<li>(unless there is reason to believe it is included in a larger package or you include a spurious non-alphanumeric character),</li>
<li><code>ml spider</code> will tell you whether it is installed or not.</li>
</ul>
</li>
<li>If the full name of a module includes <code>CUDA</code>,<ul>
<li>then the relevant <code>CUDA</code> version will typically be loaded automatically,</li>
<li>without the need to choose a CUDA-containing toolchain.</li>
</ul>
</li>
</ul>
<div class="admonition important">
<p class="admonition-title">Licenses</p>
<ul>
<li>Some specialised modules (e.g., Abaqus, Gaussian, and VASP) are license-restricted, so they may not load, or may load but refuse to run, if you are not part of the <strong>licensed</strong> user group. </li>
<li>If you run <code>ml spider</code> on a specific version of licensed software, the description may (as with VASP) or may not (as with Gaussian) specify that a license is required. </li>
<li>It is up to users to determine the licensing requirements of specialised software packages.</li>
</ul>
</div>
<h2 id="exercises__stop__recording">Exercises (Stop recording)<a class="headerlink" href="#exercises__stop__recording" title="Permanent link">&para;</a></h2>
<ul>
<li>
<p>Choose 1 or possibly 2 </p>
</li>
<li>
<p>Python bundles</p>
</li>
<li>R bundles (Tip, do not choose if using Tetralith, Alvis or Dardel)</li>
<li>Matlab and other favorite tool<ul>
<li>(Tip, try several, if you don&rsquo;t find one on your cluster)</li>
<li>Let us know if you miss something.</li>
</ul>
</li>
</ul>
<p>Break-out rooms according to interest and/or cluster</p>
<h3 id="python">Python<a class="headerlink" href="#python" title="Permanent link">&para;</a></h3>
<div class="admonition warning">
<p class="admonition-title">These clusters do NOT have Python bundles</p>
<ul>
<li>Dardel</li>
<li>Tetralith</li>
<li>Alvis</li>
</ul>
</div>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 1: Find python documentation of your cluster</p>
<details class="note">
<summary>Cluster documentation of Python</summary>
<ul>
<li><a href="https://docs.uppmax.uu.se/software/python/">UPPMAX</a></li>
<li><a href="https://docs.hpc2n.umu.se/software/apps/#python__modules">HPC2N</a></li>
<li><a href="https://lunarc-documentation.readthedocs.io/en/latest/guides/applications/Python/">LUNARC</a></li>
<li><a href="https://www.nsc.liu.se/support/">NSC</a></li>
<li><a href="https://support.pdc.kth.se/doc/applications/python/">PDC</a></li>
<li><a href="https://www.c3se.chalmers.se/documentation/module_system/python/">C3SE</a>  </li>
</ul>
</details>
</div>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 2: Find if there is a <code>matplotlib</code> package in a module</p>
<details class="note">
<summary>Tip</summary>
<p><code>ml spider</code> is good here.</p>
</details>
<details class="note">
<summary>Matplotlib across centres</summary>
<p>Matplotlib prerequisites vary significantly across HPC centres: some require none, some need one, some need more than one, and in some cases Matplotlib is only an extension of another module (<a href="https://uppmax.github.io/HPC-python/day3/new-matplotlib-intro.html#load-and-run">more info on how to find Matplotlib at different HPC centres here</a>). </p>
<p>If you only want to see what Matplotlib depends on, a good starting point is to view the output of <code>ml spider matplotlib</code> (or <code>ml avail matplotlib</code> on NSC), pick an arbitrary version, and view <code>ml spider matplotlib/&lt;version&gt;</code>.</p>
</details>
</div>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 3: Make the Python package Matplotlib available to you and test to load it in a python shell</p>
<details class="note">
<summary>Tip</summary>
<p><code>ml &lt;module name&gt;</code> is good here.</p>
</details>
<details class="note">
<summary>Example workflow</summary>
<p>All of the following code blocks in this example are taken from Cosmos.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-14-1"><a id="__codelineno-14-1" name="__codelineno-14-1" href="#__codelineno-14-1"></a>$<span class="w"> </span>ml<span class="w"> </span>spider<span class="w"> </span>matplotlib
</span><span id="__span-14-2"><a id="__codelineno-14-2" name="__codelineno-14-2" href="#__codelineno-14-2"></a>
</span><span id="__span-14-3"><a id="__codelineno-14-3" name="__codelineno-14-3" href="#__codelineno-14-3"></a>----------------------------------------------------------------------------
</span><span id="__span-14-4"><a id="__codelineno-14-4" name="__codelineno-14-4" href="#__codelineno-14-4"></a><span class="w">  </span>matplotlib:
</span><span id="__span-14-5"><a id="__codelineno-14-5" name="__codelineno-14-5" href="#__codelineno-14-5"></a>----------------------------------------------------------------------------
</span><span id="__span-14-6"><a id="__codelineno-14-6" name="__codelineno-14-6" href="#__codelineno-14-6"></a><span class="w">    </span>Description:
</span><span id="__span-14-7"><a id="__codelineno-14-7" name="__codelineno-14-7" href="#__codelineno-14-7"></a><span class="w">      </span>matplotlib<span class="w"> </span>is<span class="w"> </span>a<span class="w"> </span>python<span class="w"> </span>2D<span class="w"> </span>plotting<span class="w"> </span>library<span class="w"> </span>which<span class="w"> </span>produces<span class="w"> </span>publication
</span><span id="__span-14-8"><a id="__codelineno-14-8" name="__codelineno-14-8" href="#__codelineno-14-8"></a><span class="w">      </span>quality<span class="w"> </span>figures<span class="w"> </span><span class="k">in</span><span class="w"> </span>a<span class="w"> </span>variety<span class="w"> </span>of<span class="w"> </span>hardcopy<span class="w"> </span>formats<span class="w"> </span>and<span class="w"> </span>interactive
</span><span id="__span-14-9"><a id="__codelineno-14-9" name="__codelineno-14-9" href="#__codelineno-14-9"></a><span class="w">      </span>environments<span class="w"> </span>across<span class="w"> </span>platforms.<span class="w"> </span>matplotlib<span class="w"> </span>can<span class="w"> </span>be<span class="w"> </span>used<span class="w"> </span><span class="k">in</span><span class="w"> </span>python
</span><span id="__span-14-10"><a id="__codelineno-14-10" name="__codelineno-14-10" href="#__codelineno-14-10"></a><span class="w">      </span>scripts,<span class="w"> </span>the<span class="w"> </span>python<span class="w"> </span>and<span class="w"> </span>ipython<span class="w"> </span>shell,<span class="w"> </span>web<span class="w"> </span>application<span class="w"> </span>servers,<span class="w"> </span>and
</span><span id="__span-14-11"><a id="__codelineno-14-11" name="__codelineno-14-11" href="#__codelineno-14-11"></a><span class="w">      </span>six<span class="w"> </span>graphical<span class="w"> </span>user<span class="w"> </span>interface<span class="w"> </span>toolkits.
</span><span id="__span-14-12"><a id="__codelineno-14-12" name="__codelineno-14-12" href="#__codelineno-14-12"></a>
</span><span id="__span-14-13"><a id="__codelineno-14-13" name="__codelineno-14-13" href="#__codelineno-14-13"></a><span class="w">     </span>Versions:
</span><span id="__span-14-14"><a id="__codelineno-14-14" name="__codelineno-14-14" href="#__codelineno-14-14"></a><span class="w">        </span>matplotlib/2.2.5-Python-2.7.18
</span><span id="__span-14-15"><a id="__codelineno-14-15" name="__codelineno-14-15" href="#__codelineno-14-15"></a><span class="w">        </span>matplotlib/3.3.3
</span><span id="__span-14-16"><a id="__codelineno-14-16" name="__codelineno-14-16" href="#__codelineno-14-16"></a><span class="w">        </span>matplotlib/3.4.2
</span><span id="__span-14-17"><a id="__codelineno-14-17" name="__codelineno-14-17" href="#__codelineno-14-17"></a><span class="w">        </span>matplotlib/3.4.3
</span><span id="__span-14-18"><a id="__codelineno-14-18" name="__codelineno-14-18" href="#__codelineno-14-18"></a><span class="w">        </span>matplotlib/3.5.2
</span><span id="__span-14-19"><a id="__codelineno-14-19" name="__codelineno-14-19" href="#__codelineno-14-19"></a><span class="w">        </span>matplotlib/3.7.0
</span><span id="__span-14-20"><a id="__codelineno-14-20" name="__codelineno-14-20" href="#__codelineno-14-20"></a><span class="w">        </span>matplotlib/3.7.2
</span><span id="__span-14-21"><a id="__codelineno-14-21" name="__codelineno-14-21" href="#__codelineno-14-21"></a><span class="w">        </span>matplotlib/3.8.2
</span><span id="__span-14-22"><a id="__codelineno-14-22" name="__codelineno-14-22" href="#__codelineno-14-22"></a><span class="w">        </span>matplotlib/3.9.2
</span><span id="__span-14-23"><a id="__codelineno-14-23" name="__codelineno-14-23" href="#__codelineno-14-23"></a>
</span><span id="__span-14-24"><a id="__codelineno-14-24" name="__codelineno-14-24" href="#__codelineno-14-24"></a>----------------------------------------------------------------------------
</span></code></pre></div>
<p>If you try the above command at your local HPC centre and get a &ldquo;not found&rdquo; error, that probably means Matplotlib is an extension of another module (e.g. on Bianca, there are versions that are part of the base Python module and versions that are part of <code>python_ML_packages</code>).</p>
<p>Let us look at <code>matplotlib/3.8.2</code>, for example:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-15-1"><a id="__codelineno-15-1" name="__codelineno-15-1" href="#__codelineno-15-1"></a>$<span class="w"> </span>ml<span class="w"> </span>spider<span class="w"> </span>matplotlib/3.8.2
</span><span id="__span-15-2"><a id="__codelineno-15-2" name="__codelineno-15-2" href="#__codelineno-15-2"></a>
</span><span id="__span-15-3"><a id="__codelineno-15-3" name="__codelineno-15-3" href="#__codelineno-15-3"></a>---------------------------------------------------------------------------------
</span><span id="__span-15-4"><a id="__codelineno-15-4" name="__codelineno-15-4" href="#__codelineno-15-4"></a><span class="w">  </span>matplotlib:<span class="w"> </span>matplotlib/3.8.2
</span><span id="__span-15-5"><a id="__codelineno-15-5" name="__codelineno-15-5" href="#__codelineno-15-5"></a>---------------------------------------------------------------------------------
</span><span id="__span-15-6"><a id="__codelineno-15-6" name="__codelineno-15-6" href="#__codelineno-15-6"></a><span class="w">    </span>Description:
</span><span id="__span-15-7"><a id="__codelineno-15-7" name="__codelineno-15-7" href="#__codelineno-15-7"></a><span class="w">      </span>matplotlib<span class="w"> </span>is<span class="w"> </span>a<span class="w"> </span>python<span class="w"> </span>2D<span class="w"> </span>plotting<span class="w"> </span>library<span class="w"> </span>which<span class="w"> </span>produces<span class="w"> </span>publication
</span><span id="__span-15-8"><a id="__codelineno-15-8" name="__codelineno-15-8" href="#__codelineno-15-8"></a><span class="w">      </span>quality<span class="w"> </span>figures<span class="w"> </span><span class="k">in</span><span class="w"> </span>a<span class="w"> </span>variety<span class="w"> </span>of<span class="w"> </span>hardcopy<span class="w"> </span>formats<span class="w"> </span>and<span class="w"> </span>interactive
</span><span id="__span-15-9"><a id="__codelineno-15-9" name="__codelineno-15-9" href="#__codelineno-15-9"></a><span class="w">      </span>environments<span class="w"> </span>across<span class="w"> </span>platforms.<span class="w"> </span>matplotlib<span class="w"> </span>can<span class="w"> </span>be<span class="w"> </span>used<span class="w"> </span><span class="k">in</span><span class="w"> </span>python<span class="w"> </span>scripts,
</span><span id="__span-15-10"><a id="__codelineno-15-10" name="__codelineno-15-10" href="#__codelineno-15-10"></a><span class="w">      </span>the<span class="w"> </span>python<span class="w"> </span>and<span class="w"> </span>ipython<span class="w"> </span>shell,<span class="w"> </span>web<span class="w"> </span>application<span class="w"> </span>servers,<span class="w"> </span>and<span class="w"> </span>six<span class="w"> </span>graphical
</span><span id="__span-15-11"><a id="__codelineno-15-11" name="__codelineno-15-11" href="#__codelineno-15-11"></a><span class="w">      </span>user<span class="w"> </span>interface<span class="w"> </span>toolkits.
</span><span id="__span-15-12"><a id="__codelineno-15-12" name="__codelineno-15-12" href="#__codelineno-15-12"></a>
</span><span id="__span-15-13"><a id="__codelineno-15-13" name="__codelineno-15-13" href="#__codelineno-15-13"></a>
</span><span id="__span-15-14"><a id="__codelineno-15-14" name="__codelineno-15-14" href="#__codelineno-15-14"></a><span class="w">    </span>You<span class="w"> </span>will<span class="w"> </span>need<span class="w"> </span>to<span class="w"> </span>load<span class="w"> </span>all<span class="w"> </span>module<span class="o">(</span>s<span class="o">)</span><span class="w"> </span>on<span class="w"> </span>any<span class="w"> </span>one<span class="w"> </span>of<span class="w"> </span>the<span class="w"> </span>lines<span class="w"> </span>below<span class="w"> </span>before<span class="w"> </span>the<span class="w"> </span><span class="s2">&quot;mat</span>
</span><span id="__span-15-15"><a id="__codelineno-15-15" name="__codelineno-15-15" href="#__codelineno-15-15"></a><span class="s2">plotlib/3.8.2&quot;</span><span class="w"> </span>module<span class="w"> </span>is<span class="w"> </span>available<span class="w"> </span>to<span class="w"> </span>load.
</span><span id="__span-15-16"><a id="__codelineno-15-16" name="__codelineno-15-16" href="#__codelineno-15-16"></a>
</span><span id="__span-15-17"><a id="__codelineno-15-17" name="__codelineno-15-17" href="#__codelineno-15-17"></a><span class="w">      </span>GCC/13.2.0
</span></code></pre></div>
<p>This means that, on Cosmos at least, only GCC must be loaded before Matplotlib. However, Matplotlib is barely usable without the tools to read in or create the data arrays, so NumPy and/or Pandas are also needed. At most facilities, that means SciPy-bundle is required.</p>
<p>Note that <code>ml show matplotlib/&lt;version&gt;</code> does <strong>not</strong> show which Python version is associated with that version of Matplotlib. If <code>GCC</code> is loaded, then you can use <code>ml avail</code> with <code>Python</code>, <code>matplotlib</code>, and/or <code>SciPy-bundle</code> to see which versions of these are available to load.</p>
<p>The more typical scenario is that you want to move code developed on a personal laptop to the cluster. Then you will mainly be constrained to a range of Python versions <code>Python/X.Y.Z</code>, in which X absolutely <em>must</em> match what you used, Y <em>should</em> match but may be flexible by one or two versions, and Z is usually not that important. In a bash terminal, you can check your Python version with <code>python --version</code>.</p>
<p>Let&rsquo;s say you built a script using Python 3.11.8 and a compatible version of Matplotlib on your own laptop. Glob patterns do not work to select subsets of <code>ml spider</code> or <code>ml avail</code> outputs, so one must view the full list with <code>ml spider Python</code> (<code>ml spider cray-python</code> on Dardel). Here is the output on Cosmos:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-16-1"><a id="__codelineno-16-1" name="__codelineno-16-1" href="#__codelineno-16-1"></a>$<span class="w"> </span>ml<span class="w"> </span>spider<span class="w"> </span>Python
</span><span id="__span-16-2"><a id="__codelineno-16-2" name="__codelineno-16-2" href="#__codelineno-16-2"></a>---------------------------------------------------------------------------------
</span><span id="__span-16-3"><a id="__codelineno-16-3" name="__codelineno-16-3" href="#__codelineno-16-3"></a><span class="w">  </span>Python:
</span><span id="__span-16-4"><a id="__codelineno-16-4" name="__codelineno-16-4" href="#__codelineno-16-4"></a>---------------------------------------------------------------------------------
</span><span id="__span-16-5"><a id="__codelineno-16-5" name="__codelineno-16-5" href="#__codelineno-16-5"></a><span class="w">    </span>Description:
</span><span id="__span-16-6"><a id="__codelineno-16-6" name="__codelineno-16-6" href="#__codelineno-16-6"></a><span class="w">      </span>Python<span class="w"> </span>is<span class="w"> </span>a<span class="w"> </span>programming<span class="w"> </span>language<span class="w"> </span>that<span class="w"> </span>lets<span class="w"> </span>you<span class="w"> </span>work<span class="w"> </span>more<span class="w"> </span>quickly<span class="w"> </span>and<span class="w"> </span>integrate
</span><span id="__span-16-7"><a id="__codelineno-16-7" name="__codelineno-16-7" href="#__codelineno-16-7"></a>your<span class="w"> </span>systems<span class="w"> </span>more<span class="w"> </span>effectively.
</span><span id="__span-16-8"><a id="__codelineno-16-8" name="__codelineno-16-8" href="#__codelineno-16-8"></a>
</span><span id="__span-16-9"><a id="__codelineno-16-9" name="__codelineno-16-9" href="#__codelineno-16-9"></a><span class="w">     </span>Versions:
</span><span id="__span-16-10"><a id="__codelineno-16-10" name="__codelineno-16-10" href="#__codelineno-16-10"></a><span class="w">        </span>Python/2.7.18-bare
</span><span id="__span-16-11"><a id="__codelineno-16-11" name="__codelineno-16-11" href="#__codelineno-16-11"></a><span class="w">        </span>Python/2.7.18
</span><span id="__span-16-12"><a id="__codelineno-16-12" name="__codelineno-16-12" href="#__codelineno-16-12"></a><span class="w">        </span>Python/3.8.6
</span><span id="__span-16-13"><a id="__codelineno-16-13" name="__codelineno-16-13" href="#__codelineno-16-13"></a><span class="w">        </span>Python/3.9.5-bare
</span><span id="__span-16-14"><a id="__codelineno-16-14" name="__codelineno-16-14" href="#__codelineno-16-14"></a><span class="w">        </span>Python/3.9.5
</span><span id="__span-16-15"><a id="__codelineno-16-15" name="__codelineno-16-15" href="#__codelineno-16-15"></a><span class="w">        </span>Python/3.9.6-bare
</span><span id="__span-16-16"><a id="__codelineno-16-16" name="__codelineno-16-16" href="#__codelineno-16-16"></a><span class="w">        </span>Python/3.9.6
</span><span id="__span-16-17"><a id="__codelineno-16-17" name="__codelineno-16-17" href="#__codelineno-16-17"></a><span class="w">        </span>Python/3.10.4-bare
</span><span id="__span-16-18"><a id="__codelineno-16-18" name="__codelineno-16-18" href="#__codelineno-16-18"></a><span class="w">        </span>Python/3.10.4
</span><span id="__span-16-19"><a id="__codelineno-16-19" name="__codelineno-16-19" href="#__codelineno-16-19"></a><span class="w">        </span>Python/3.10.8-bare
</span><span id="__span-16-20"><a id="__codelineno-16-20" name="__codelineno-16-20" href="#__codelineno-16-20"></a><span class="w">        </span>Python/3.10.8
</span><span id="__span-16-21"><a id="__codelineno-16-21" name="__codelineno-16-21" href="#__codelineno-16-21"></a><span class="w">        </span>Python/3.11.3
</span><span id="__span-16-22"><a id="__codelineno-16-22" name="__codelineno-16-22" href="#__codelineno-16-22"></a><span class="w">        </span>Python/3.11.5
</span><span id="__span-16-23"><a id="__codelineno-16-23" name="__codelineno-16-23" href="#__codelineno-16-23"></a><span class="w">        </span>Python/3.12.3
</span><span id="__span-16-24"><a id="__codelineno-16-24" name="__codelineno-16-24" href="#__codelineno-16-24"></a><span class="w">     </span>Other<span class="w"> </span>possible<span class="w"> </span>modules<span class="w"> </span>matches:
</span><span id="__span-16-25"><a id="__codelineno-16-25" name="__codelineno-16-25" href="#__codelineno-16-25"></a><span class="w">        </span>Biopython<span class="w">  </span>GitPython<span class="w">  </span>IPython<span class="w">  </span>Python-bundle<span class="w">  </span>Python-bundle-PyPI
</span><span id="__span-16-26"><a id="__codelineno-16-26" name="__codelineno-16-26" href="#__codelineno-16-26"></a><span class="w">        </span>bx-python<span class="w">  </span>flatbuffers-python<span class="w">  </span>graphviz-python<span class="w">  </span>meson-python<span class="w">  </span>...
</span></code></pre></div>
<p>The closest result is <code>Python/3.11.5</code> (though probably anything from 3.10.x to 3.12.x would work). Let&rsquo;s check what that requires:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-17-1"><a id="__codelineno-17-1" name="__codelineno-17-1" href="#__codelineno-17-1"></a>$<span class="w"> </span>ml<span class="w"> </span>spider<span class="w"> </span>Python/3.11.5
</span><span id="__span-17-2"><a id="__codelineno-17-2" name="__codelineno-17-2" href="#__codelineno-17-2"></a>---------------------------------------------------------------------------------
</span><span id="__span-17-3"><a id="__codelineno-17-3" name="__codelineno-17-3" href="#__codelineno-17-3"></a><span class="w">  </span>Python:<span class="w"> </span>Python/3.11.5
</span><span id="__span-17-4"><a id="__codelineno-17-4" name="__codelineno-17-4" href="#__codelineno-17-4"></a>---------------------------------------------------------------------------------
</span><span id="__span-17-5"><a id="__codelineno-17-5" name="__codelineno-17-5" href="#__codelineno-17-5"></a><span class="w">    </span>Description:
</span><span id="__span-17-6"><a id="__codelineno-17-6" name="__codelineno-17-6" href="#__codelineno-17-6"></a><span class="w">      </span>Python<span class="w"> </span>is<span class="w"> </span>a<span class="w"> </span>programming<span class="w"> </span>language<span class="w"> </span>that<span class="w"> </span>lets<span class="w"> </span>you<span class="w"> </span>work<span class="w"> </span>more<span class="w"> </span>quickly<span class="w"> </span>and
</span><span id="__span-17-7"><a id="__codelineno-17-7" name="__codelineno-17-7" href="#__codelineno-17-7"></a><span class="w">      </span>integrate<span class="w"> </span>your<span class="w"> </span>systems<span class="w"> </span>more<span class="w"> </span>effectively.
</span><span id="__span-17-8"><a id="__codelineno-17-8" name="__codelineno-17-8" href="#__codelineno-17-8"></a>
</span><span id="__span-17-9"><a id="__codelineno-17-9" name="__codelineno-17-9" href="#__codelineno-17-9"></a>
</span><span id="__span-17-10"><a id="__codelineno-17-10" name="__codelineno-17-10" href="#__codelineno-17-10"></a><span class="w">    </span>You<span class="w"> </span>will<span class="w"> </span>need<span class="w"> </span>to<span class="w"> </span>load<span class="w"> </span>all<span class="w"> </span>module<span class="o">(</span>s<span class="o">)</span><span class="w"> </span>on<span class="w"> </span>any<span class="w"> </span>one<span class="w"> </span>of<span class="w"> </span>the<span class="w"> </span>lines<span class="w"> </span>below<span class="w"> </span>before<span class="w"> </span>the
</span><span id="__span-17-11"><a id="__codelineno-17-11" name="__codelineno-17-11" href="#__codelineno-17-11"></a><span class="w">    </span><span class="s2">&quot;Python/3.11.5&quot;</span><span class="w"> </span>module<span class="w"> </span>is<span class="w"> </span>available<span class="w"> </span>to<span class="w"> </span>load.
</span><span id="__span-17-12"><a id="__codelineno-17-12" name="__codelineno-17-12" href="#__codelineno-17-12"></a>
</span><span id="__span-17-13"><a id="__codelineno-17-13" name="__codelineno-17-13" href="#__codelineno-17-13"></a><span class="w">      </span>GCCcore/13.2.0
</span><span id="__span-17-14"><a id="__codelineno-17-14" name="__codelineno-17-14" href="#__codelineno-17-14"></a>
</span><span id="__span-17-15"><a id="__codelineno-17-15" name="__codelineno-17-15" href="#__codelineno-17-15"></a><span class="w">    </span>Help:
</span><span id="__span-17-16"><a id="__codelineno-17-16" name="__codelineno-17-16" href="#__codelineno-17-16"></a>
</span><span id="__span-17-17"><a id="__codelineno-17-17" name="__codelineno-17-17" href="#__codelineno-17-17"></a><span class="w">      </span><span class="nv">Description</span>
</span><span id="__span-17-18"><a id="__codelineno-17-18" name="__codelineno-17-18" href="#__codelineno-17-18"></a><span class="w">      </span><span class="o">===========</span>
</span><span id="__span-17-19"><a id="__codelineno-17-19" name="__codelineno-17-19" href="#__codelineno-17-19"></a><span class="w">      </span>Python<span class="w"> </span>is<span class="w"> </span>a<span class="w"> </span>programming<span class="w"> </span>language<span class="w"> </span>that<span class="w"> </span>lets<span class="w"> </span>you<span class="w"> </span>work<span class="w"> </span>more<span class="w"> </span>quickly<span class="w"> </span>and<span class="w"> </span>integrate<span class="w"> </span>
</span><span id="__span-17-20"><a id="__codelineno-17-20" name="__codelineno-17-20" href="#__codelineno-17-20"></a>your<span class="w"> </span>systems<span class="w"> </span>more<span class="w"> </span>effectively.
</span><span id="__span-17-21"><a id="__codelineno-17-21" name="__codelineno-17-21" href="#__codelineno-17-21"></a>
</span><span id="__span-17-22"><a id="__codelineno-17-22" name="__codelineno-17-22" href="#__codelineno-17-22"></a>
</span><span id="__span-17-23"><a id="__codelineno-17-23" name="__codelineno-17-23" href="#__codelineno-17-23"></a><span class="w">      </span>More<span class="w"> </span><span class="nv">information</span>
</span><span id="__span-17-24"><a id="__codelineno-17-24" name="__codelineno-17-24" href="#__codelineno-17-24"></a><span class="w">      </span><span class="o">================</span>
</span><span id="__span-17-25"><a id="__codelineno-17-25" name="__codelineno-17-25" href="#__codelineno-17-25"></a><span class="w">       </span>-<span class="w"> </span>Homepage:<span class="w"> </span>https://python.org/
</span><span id="__span-17-26"><a id="__codelineno-17-26" name="__codelineno-17-26" href="#__codelineno-17-26"></a>
</span><span id="__span-17-27"><a id="__codelineno-17-27" name="__codelineno-17-27" href="#__codelineno-17-27"></a>
</span><span id="__span-17-28"><a id="__codelineno-17-28" name="__codelineno-17-28" href="#__codelineno-17-28"></a><span class="w">      </span>Included<span class="w"> </span><span class="nv">extensions</span>
</span><span id="__span-17-29"><a id="__codelineno-17-29" name="__codelineno-17-29" href="#__codelineno-17-29"></a><span class="w">      </span><span class="o">===================</span>
</span><span id="__span-17-30"><a id="__codelineno-17-30" name="__codelineno-17-30" href="#__codelineno-17-30"></a><span class="w">      </span>flit_core-3.9.0,<span class="w"> </span>packaging-23.2,<span class="w"> </span>pip-23.2.1,<span class="w"> </span>setuptools-68.2.2,<span class="w"> </span>setuptools-
</span><span id="__span-17-31"><a id="__codelineno-17-31" name="__codelineno-17-31" href="#__codelineno-17-31"></a><span class="w">      </span>scm-8.0.4,<span class="w"> </span>tomli-2.0.1,<span class="w"> </span>typing_extensions-4.8.0,<span class="w"> </span>wheel-0.41.2
</span></code></pre></div>
<p>On this cluster, the base Python module requires GCCcore, but we already saw that Matplotlib requires GCC (whch GCCcore is part of). In fact, nearly every other Python-based module apart from the bare Python itself requires GCC, so you may as well use GCC every time.</p>
<p>On some facilities, each version of Matplotlib and SciPy-bundle is only be associated with one Python version, so you can load them all at once, using the GCC version to select for everything else, like this:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-18-1"><a id="__codelineno-18-1" name="__codelineno-18-1" href="#__codelineno-18-1"></a>ml<span class="w"> </span>GCC/13.2.0<span class="w"> </span>Python<span class="w"> </span>matplotlib<span class="w"> </span>SciPy-bundle
</span></code></pre></div>
<p>However, this is considered bad practice since sometimes additional versions are installed later. We should instead check <code>ml avail</code> to see what versions of Matplotlib and Scipy-bundle we can load:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-19-1"><a id="__codelineno-19-1" name="__codelineno-19-1" href="#__codelineno-19-1"></a>$<span class="w"> </span>ml<span class="w"> </span>avail<span class="w"> </span>Scipy-bundle
</span><span id="__span-19-2"><a id="__codelineno-19-2" name="__codelineno-19-2" href="#__codelineno-19-2"></a>
</span><span id="__span-19-3"><a id="__codelineno-19-3" name="__codelineno-19-3" href="#__codelineno-19-3"></a>-----------------------------<span class="w"> </span>/sw/easybuild_milan/modules/all/Compiler/GCC/13.2.0<span class="w"> </span>-----------------------------
</span><span id="__span-19-4"><a id="__codelineno-19-4" name="__codelineno-19-4" href="#__codelineno-19-4"></a><span class="w">   </span>SciPy-bundle/2023.11
</span></code></pre></div>
<p>(Note: some output omitted for brevity)</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-20-1"><a id="__codelineno-20-1" name="__codelineno-20-1" href="#__codelineno-20-1"></a>$<span class="w"> </span>ml<span class="w"> </span>avail<span class="w"> </span>Matplotlib
</span><span id="__span-20-2"><a id="__codelineno-20-2" name="__codelineno-20-2" href="#__codelineno-20-2"></a>
</span><span id="__span-20-3"><a id="__codelineno-20-3" name="__codelineno-20-3" href="#__codelineno-20-3"></a>-----------------------------<span class="w"> </span>/sw/easybuild_milan/modules/all/Compiler/GCC/13.2.0<span class="w"> </span>-----------------------------
</span><span id="__span-20-4"><a id="__codelineno-20-4" name="__codelineno-20-4" href="#__codelineno-20-4"></a><span class="w">   </span>matplotlib/3.8.2
</span></code></pre></div>
<p>Then the one-line loading command should look like this:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-21-1"><a id="__codelineno-21-1" name="__codelineno-21-1" href="#__codelineno-21-1"></a>ml<span class="w"> </span>GCC/13.2.0<span class="w"> </span>Python/3.11.5<span class="w"> </span>matplotlib/3.8.2<span class="w"> </span>SciPy-bundle/2023.11
</span></code></pre></div>
<p>If we check what was loaded with <code>ml</code> or <code>module list</code>, the output looks like this: </p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-22-1"><a id="__codelineno-22-1" name="__codelineno-22-1" href="#__codelineno-22-1"></a>$<span class="w"> </span>ml
</span><span id="__span-22-2"><a id="__codelineno-22-2" name="__codelineno-22-2" href="#__codelineno-22-2"></a>
</span><span id="__span-22-3"><a id="__codelineno-22-3" name="__codelineno-22-3" href="#__codelineno-22-3"></a>Currently<span class="w"> </span>Loaded<span class="w"> </span>Modules:
</span><span id="__span-22-4"><a id="__codelineno-22-4" name="__codelineno-22-4" href="#__codelineno-22-4"></a><span class="w">  </span><span class="m">1</span><span class="o">)</span><span class="w"> </span>SoftwareTree/Milan<span class="w">         </span><span class="o">(</span>S<span class="o">)</span><span class="w">  </span><span class="m">26</span><span class="o">)</span><span class="w"> </span>expat/2.5.0
</span><span id="__span-22-5"><a id="__codelineno-22-5" name="__codelineno-22-5" href="#__codelineno-22-5"></a><span class="w">  </span><span class="m">2</span><span class="o">)</span><span class="w"> </span>GCCcore/13.2.0<span class="w">                  </span><span class="m">27</span><span class="o">)</span><span class="w"> </span>util-linux/2.39
</span><span id="__span-22-6"><a id="__codelineno-22-6" name="__codelineno-22-6" href="#__codelineno-22-6"></a><span class="w">  </span><span class="m">3</span><span class="o">)</span><span class="w"> </span>zlib/1.2.13<span class="w">                     </span><span class="m">28</span><span class="o">)</span><span class="w"> </span>fontconfig/2.14.2
</span><span id="__span-22-7"><a id="__codelineno-22-7" name="__codelineno-22-7" href="#__codelineno-22-7"></a><span class="w">  </span><span class="m">4</span><span class="o">)</span><span class="w"> </span>binutils/2.40<span class="w">                   </span><span class="m">29</span><span class="o">)</span><span class="w"> </span>xorg-macros/1.20.0
</span><span id="__span-22-8"><a id="__codelineno-22-8" name="__codelineno-22-8" href="#__codelineno-22-8"></a><span class="w">  </span><span class="m">5</span><span class="o">)</span><span class="w"> </span>GCC/13.2.0<span class="w">                      </span><span class="m">30</span><span class="o">)</span><span class="w"> </span>libpciaccess/0.17
</span><span id="__span-22-9"><a id="__codelineno-22-9" name="__codelineno-22-9" href="#__codelineno-22-9"></a><span class="w">  </span><span class="m">6</span><span class="o">)</span><span class="w"> </span>bzip2/1.0.8<span class="w">                     </span><span class="m">31</span><span class="o">)</span><span class="w"> </span>X11/20231019
</span><span id="__span-22-10"><a id="__codelineno-22-10" name="__codelineno-22-10" href="#__codelineno-22-10"></a><span class="w">  </span><span class="m">7</span><span class="o">)</span><span class="w"> </span>ncurses/6.4<span class="w">                     </span><span class="m">32</span><span class="o">)</span><span class="w"> </span>Tk/8.6.13
</span><span id="__span-22-11"><a id="__codelineno-22-11" name="__codelineno-22-11" href="#__codelineno-22-11"></a><span class="w">  </span><span class="m">8</span><span class="o">)</span><span class="w"> </span>libreadline/8.2<span class="w">                 </span><span class="m">33</span><span class="o">)</span><span class="w"> </span>Tkinter/3.11.5
</span><span id="__span-22-12"><a id="__codelineno-22-12" name="__codelineno-22-12" href="#__codelineno-22-12"></a><span class="w">  </span><span class="m">9</span><span class="o">)</span><span class="w"> </span>Tcl/8.6.13<span class="w">                      </span><span class="m">34</span><span class="o">)</span><span class="w"> </span>NASM/2.16.01
</span><span id="__span-22-13"><a id="__codelineno-22-13" name="__codelineno-22-13" href="#__codelineno-22-13"></a><span class="w"> </span><span class="m">10</span><span class="o">)</span><span class="w"> </span>SQLite/3.43.1<span class="w">                   </span><span class="m">35</span><span class="o">)</span><span class="w"> </span>libjpeg-turbo/3.0.1
</span><span id="__span-22-14"><a id="__codelineno-22-14" name="__codelineno-22-14" href="#__codelineno-22-14"></a><span class="w"> </span><span class="m">11</span><span class="o">)</span><span class="w"> </span>XZ/5.4.4<span class="w">                        </span><span class="m">36</span><span class="o">)</span><span class="w"> </span>jbigkit/2.1
</span><span id="__span-22-15"><a id="__codelineno-22-15" name="__codelineno-22-15" href="#__codelineno-22-15"></a><span class="w"> </span><span class="m">12</span><span class="o">)</span><span class="w"> </span>libffi/3.4.4<span class="w">                    </span><span class="m">37</span><span class="o">)</span><span class="w"> </span>gzip/1.13
</span><span id="__span-22-16"><a id="__codelineno-22-16" name="__codelineno-22-16" href="#__codelineno-22-16"></a><span class="w"> </span><span class="m">13</span><span class="o">)</span><span class="w"> </span>OpenSSL/1.1<span class="w">                     </span><span class="m">38</span><span class="o">)</span><span class="w"> </span>lz4/1.9.4
</span><span id="__span-22-17"><a id="__codelineno-22-17" name="__codelineno-22-17" href="#__codelineno-22-17"></a><span class="w"> </span><span class="m">14</span><span class="o">)</span><span class="w"> </span>Python/3.11.5<span class="w">                   </span><span class="m">39</span><span class="o">)</span><span class="w"> </span>zstd/1.5.5
</span><span id="__span-22-18"><a id="__codelineno-22-18" name="__codelineno-22-18" href="#__codelineno-22-18"></a><span class="w"> </span><span class="m">15</span><span class="o">)</span><span class="w"> </span>OpenBLAS/0.3.24<span class="w">                 </span><span class="m">40</span><span class="o">)</span><span class="w"> </span>libdeflate/1.19
</span><span id="__span-22-19"><a id="__codelineno-22-19" name="__codelineno-22-19" href="#__codelineno-22-19"></a><span class="w"> </span><span class="m">16</span><span class="o">)</span><span class="w"> </span>FlexiBLAS/3.3.1<span class="w">                 </span><span class="m">41</span><span class="o">)</span><span class="w"> </span>LibTIFF/4.6.0
</span><span id="__span-22-20"><a id="__codelineno-22-20" name="__codelineno-22-20" href="#__codelineno-22-20"></a><span class="w"> </span><span class="m">17</span><span class="o">)</span><span class="w"> </span>FFTW/3.3.10<span class="w">                     </span><span class="m">42</span><span class="o">)</span><span class="w"> </span>giflib/5.2.1
</span><span id="__span-22-21"><a id="__codelineno-22-21" name="__codelineno-22-21" href="#__codelineno-22-21"></a><span class="w"> </span><span class="m">18</span><span class="o">)</span><span class="w"> </span>cffi/1.15.1<span class="w">                     </span><span class="m">43</span><span class="o">)</span><span class="w"> </span>libwebp/1.3.2
</span><span id="__span-22-22"><a id="__codelineno-22-22" name="__codelineno-22-22" href="#__codelineno-22-22"></a><span class="w"> </span><span class="m">19</span><span class="o">)</span><span class="w"> </span>cryptography/41.0.5<span class="w">             </span><span class="m">44</span><span class="o">)</span><span class="w"> </span>OpenJPEG/2.5.0
</span><span id="__span-22-23"><a id="__codelineno-22-23" name="__codelineno-22-23" href="#__codelineno-22-23"></a><span class="w"> </span><span class="m">20</span><span class="o">)</span><span class="w"> </span>virtualenv/20.24.6<span class="w">              </span><span class="m">45</span><span class="o">)</span><span class="w"> </span>LittleCMS/2.15
</span><span id="__span-22-24"><a id="__codelineno-22-24" name="__codelineno-22-24" href="#__codelineno-22-24"></a><span class="w"> </span><span class="m">21</span><span class="o">)</span><span class="w"> </span>Python-bundle-PyPI/2023.10<span class="w">      </span><span class="m">46</span><span class="o">)</span><span class="w"> </span>Pillow/10.2.0
</span><span id="__span-22-25"><a id="__codelineno-22-25" name="__codelineno-22-25" href="#__codelineno-22-25"></a><span class="w"> </span><span class="m">22</span><span class="o">)</span><span class="w"> </span>pybind11/2.11.1<span class="w">                 </span><span class="m">47</span><span class="o">)</span><span class="w"> </span>Qhull/2020.2
</span><span id="__span-22-26"><a id="__codelineno-22-26" name="__codelineno-22-26" href="#__codelineno-22-26"></a><span class="w"> </span><span class="m">23</span><span class="o">)</span><span class="w"> </span>libpng/1.6.40<span class="w">                   </span><span class="m">48</span><span class="o">)</span><span class="w"> </span>matplotlib/3.8.2
</span><span id="__span-22-27"><a id="__codelineno-22-27" name="__codelineno-22-27" href="#__codelineno-22-27"></a><span class="w"> </span><span class="m">24</span><span class="o">)</span><span class="w"> </span>Brotli/1.1.0<span class="w">                    </span><span class="m">49</span><span class="o">)</span><span class="w"> </span>SciPy-bundle/2023.11
</span><span id="__span-22-28"><a id="__codelineno-22-28" name="__codelineno-22-28" href="#__codelineno-22-28"></a><span class="w"> </span><span class="m">25</span><span class="o">)</span><span class="w"> </span>freetype/2.13.2
</span><span id="__span-22-29"><a id="__codelineno-22-29" name="__codelineno-22-29" href="#__codelineno-22-29"></a>
</span><span id="__span-22-30"><a id="__codelineno-22-30" name="__codelineno-22-30" href="#__codelineno-22-30"></a><span class="w">  </span>Where:
</span><span id="__span-22-31"><a id="__codelineno-22-31" name="__codelineno-22-31" href="#__codelineno-22-31"></a><span class="w">   </span>S:<span class="w">  </span>Module<span class="w"> </span>is<span class="w"> </span>Sticky,<span class="w"> </span>requires<span class="w"> </span>--force<span class="w"> </span>to<span class="w"> </span>unload<span class="w"> </span>or<span class="w"> </span>purge
</span></code></pre></div>
<p>If you are comfortable editing code in a basic text editor and running at the command-line, the modules used in the example above are all you need. For more information on choosing and loading IDEs to work with Matplotlib graphics interactively, we refer readers to <a href="https://uppmax.github.io/HPC-python/day2/IDEs.html">this documentation from the Python for HPC course</a>.</p>
</details>
</div>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 4: Check if the python package XX is available in the present environment</p>
<p>Extensions can be hard to find without knowing what includes them, but it is easy to check if modules that are already loaded added the extension silently. If you cannot find a package you want with <code>ml avail</code>, <code>ml spider</code>, or <code>ml show &lt;module&gt;</code>, you should also check <code>pip list</code> and <code>grep</code> for the package after loading the rest of your modules. </p>
<details class="note">
<summary>Example: <code>psutil</code></summary>
<p>For example, <code>psutil</code> is part of Python-bundle-PyPI, which is silently loaded with any SciPy-bundle. Here is the easiest way to find <code>psutil</code>:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-23-1"><a id="__codelineno-23-1" name="__codelineno-23-1" href="#__codelineno-23-1"></a>$<span class="w"> </span>pip<span class="w"> </span>list<span class="w"> </span><span class="p">|</span><span class="w"> </span>grep<span class="w"> </span>psutil
</span><span id="__span-23-2"><a id="__codelineno-23-2" name="__codelineno-23-2" href="#__codelineno-23-2"></a>psutil<span class="w">                            </span><span class="m">5</span>.9.5
</span><span id="__span-23-3"><a id="__codelineno-23-3" name="__codelineno-23-3" href="#__codelineno-23-3"></a>
</span><span id="__span-23-4"><a id="__codelineno-23-4" name="__codelineno-23-4" href="#__codelineno-23-4"></a><span class="o">[</span>notice<span class="o">]</span><span class="w"> </span>A<span class="w"> </span>new<span class="w"> </span>release<span class="w"> </span>of<span class="w"> </span>pip<span class="w"> </span>is<span class="w"> </span>available:<span class="w"> </span><span class="m">23</span>.1.2<span class="w"> </span>-&gt;<span class="w"> </span><span class="m">25</span>.1.1
</span><span id="__span-23-5"><a id="__codelineno-23-5" name="__codelineno-23-5" href="#__codelineno-23-5"></a><span class="o">[</span>notice<span class="o">]</span><span class="w"> </span>To<span class="w"> </span>update,<span class="w"> </span>run:<span class="w"> </span>pip<span class="w"> </span>install<span class="w"> </span>--upgrade<span class="w"> </span>pip
</span></code></pre></div>
<p>The <code>pip list | grep</code> approach is also helpful if you want to see the version of a package without having to open a Python interpreter.</p>
</details>
<div class="admonition note">
<p class="admonition-title">Tip</p>
<p>The same list (and grep) approach works for Anaconda3. The only difference is that you should use <code>conda list</code> instead of <code>pip list</code> (although <code>pip list</code> usually still works). The Anaconda3 module file does <strong>not</strong> list the included extensions, so <code>conda list | grep &lt;package&gt;</code> is also the <em>only</em> way to see if a package is included without starting up a Python command line interface.</p>
</div>
</div>
<h3 id="r_1">R<a class="headerlink" href="#r_1" title="Permanent link">&para;</a></h3>
<div class="admonition warning">
<p class="admonition-title">These clusters do NOT have R bundle</p>
<ul>
<li>Bianca has R_packages that contains CRAN and Bioconductor packages<ul>
<li>You can load that instead of a bundle in the below exercises.</li>
</ul>
</li>
<li>Dardel</li>
<li>Tetralith</li>
<li>Alvis</li>
</ul>
</div>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 1: Find R documentation of your cluster</p>
<details class="note">
<summary>Tip</summary>
<ul>
<li><a href="https://www.c3se.chalmers.se/documentation/software/development/r/">C3SE</a></li>
<li><a href="https://www.hpc2n.umu.se/resources/software/r">HPC2N</a></li>
<li><a href="https://lunarc-documentation.readthedocs.io/en/latest/guides/applications/Rscript/">LUNARC</a></li>
<li><a href="https://ipv4.www.nsc.liu.se/software/catalogue/tetralith/modules/r.html">NSC</a></li>
<li><a href="https://support.pdc.kth.se/doc/contact/contact_support/">PDC</a></li>
<li><a href="https://docs.uppmax.uu.se/software/r.md">UPPMAX</a></li>
</ul>
</details>
</div>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 2: Load R and start it!</p>
<details class="note">
<summary>Answer</summary>
<p>Course page <a href="https://uppmax.github.io/R-matlab-julia-HPC/r/load_run/">how to load</a></p>
</details>
</div>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 3: Make the R package <code>Seurat</code> available to you by loading Bioconductor and test to load it (<code>library(Seurat)</code>) in a R shell</p>
<details class="note">
<summary>Example at Cosmos</summary>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-24-1"><a id="__codelineno-24-1" name="__codelineno-24-1" href="#__codelineno-24-1"></a>$<span class="w"> </span>ml<span class="w"> </span>spider<span class="w"> </span>bioconductor
</span><span id="__span-24-2"><a id="__codelineno-24-2" name="__codelineno-24-2" href="#__codelineno-24-2"></a>
</span><span id="__span-24-3"><a id="__codelineno-24-3" name="__codelineno-24-3" href="#__codelineno-24-3"></a>---------------------------------------------------------------------------------
</span><span id="__span-24-4"><a id="__codelineno-24-4" name="__codelineno-24-4" href="#__codelineno-24-4"></a><span class="w">  </span>R-bundle-Bioconductor:
</span><span id="__span-24-5"><a id="__codelineno-24-5" name="__codelineno-24-5" href="#__codelineno-24-5"></a>---------------------------------------------------------------------------------
</span><span id="__span-24-6"><a id="__codelineno-24-6" name="__codelineno-24-6" href="#__codelineno-24-6"></a><span class="w">    </span>Description:
</span><span id="__span-24-7"><a id="__codelineno-24-7" name="__codelineno-24-7" href="#__codelineno-24-7"></a><span class="w">      </span>Bioconductor<span class="w"> </span>provides<span class="w"> </span>tools<span class="w"> </span><span class="k">for</span><span class="w"> </span>the<span class="w"> </span>analysis<span class="w"> </span>and<span class="w"> </span>coprehension<span class="w"> </span>of
</span><span id="__span-24-8"><a id="__codelineno-24-8" name="__codelineno-24-8" href="#__codelineno-24-8"></a><span class="w">      </span>high-throughput<span class="w"> </span>genomic<span class="w"> </span>data.
</span><span id="__span-24-9"><a id="__codelineno-24-9" name="__codelineno-24-9" href="#__codelineno-24-9"></a>
</span><span id="__span-24-10"><a id="__codelineno-24-10" name="__codelineno-24-10" href="#__codelineno-24-10"></a><span class="w">     </span>Versions:
</span><span id="__span-24-11"><a id="__codelineno-24-11" name="__codelineno-24-11" href="#__codelineno-24-11"></a><span class="w">        </span>R-bundle-Bioconductor/3.15-R-4.2.1
</span><span id="__span-24-12"><a id="__codelineno-24-12" name="__codelineno-24-12" href="#__codelineno-24-12"></a><span class="w">        </span>R-bundle-Bioconductor/3.18-R-4.3.2
</span><span id="__span-24-13"><a id="__codelineno-24-13" name="__codelineno-24-13" href="#__codelineno-24-13"></a><span class="w">        </span>R-bundle-Bioconductor/3.18-R-4.4.1
</span><span id="__span-24-14"><a id="__codelineno-24-14" name="__codelineno-24-14" href="#__codelineno-24-14"></a><span class="w">        </span>R-bundle-Bioconductor/3.19-R-4.4.1
</span><span id="__span-24-15"><a id="__codelineno-24-15" name="__codelineno-24-15" href="#__codelineno-24-15"></a>
</span><span id="__span-24-16"><a id="__codelineno-24-16" name="__codelineno-24-16" href="#__codelineno-24-16"></a>---------------------------------------------------------------------------------
</span><span id="__span-24-17"><a id="__codelineno-24-17" name="__codelineno-24-17" href="#__codelineno-24-17"></a><span class="w">  </span>For<span class="w"> </span>detailed<span class="w"> </span>information<span class="w"> </span>about<span class="w"> </span>a<span class="w"> </span>specific<span class="w"> </span><span class="s2">&quot;R-bundle-Bioconductor&quot;</span><span class="w"> </span>package<span class="w"> </span><span class="o">(</span>includin
</span><span id="__span-24-18"><a id="__codelineno-24-18" name="__codelineno-24-18" href="#__codelineno-24-18"></a>g<span class="w"> </span>how<span class="w"> </span>to<span class="w"> </span>load<span class="w"> </span>the<span class="w"> </span>modules<span class="o">)</span><span class="w"> </span>use<span class="w"> </span>the<span class="w"> </span>module<span class="err">&#39;</span>s<span class="w"> </span>full<span class="w"> </span>name.
</span><span id="__span-24-19"><a id="__codelineno-24-19" name="__codelineno-24-19" href="#__codelineno-24-19"></a><span class="w">  </span>Note<span class="w"> </span>that<span class="w"> </span>names<span class="w"> </span>that<span class="w"> </span>have<span class="w"> </span>a<span class="w"> </span>trailing<span class="w"> </span><span class="o">(</span>E<span class="o">)</span><span class="w"> </span>are<span class="w"> </span>extensions<span class="w"> </span>provided<span class="w"> </span>by<span class="w"> </span>other<span class="w"> </span>modules.
</span><span id="__span-24-20"><a id="__codelineno-24-20" name="__codelineno-24-20" href="#__codelineno-24-20"></a><span class="w">  </span>For<span class="w"> </span>example:
</span><span id="__span-24-21"><a id="__codelineno-24-21" name="__codelineno-24-21" href="#__codelineno-24-21"></a>
</span><span id="__span-24-22"><a id="__codelineno-24-22" name="__codelineno-24-22" href="#__codelineno-24-22"></a><span class="w">     </span>$<span class="w"> </span>module<span class="w"> </span>spider<span class="w"> </span>R-bundle-Bioconductor/3.19-R-4.4.1
</span><span id="__span-24-23"><a id="__codelineno-24-23" name="__codelineno-24-23" href="#__codelineno-24-23"></a>---------------------------------------------------------------------------------
</span></code></pre></div>
<p>Notice that in this case, there are 2 versions of the Bioconductor bundle associated with <code>R/4.4.1</code>, and that there are 2 versions of R associated with <code>R-bundle-Bioconductor/3.18</code>. Do not rely on the prerequisites to set which version of <code>R-bundle-Bioconductor</code> gets loaded.</p>
<p>To check the prerequisites with <code>ml spider</code>, the specific version number must be included anyway, for all software.</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-25-1"><a id="__codelineno-25-1" name="__codelineno-25-1" href="#__codelineno-25-1"></a>$<span class="w"> </span>ml<span class="w"> </span>spider<span class="w"> </span>R-bundle-Bioconductor/3.18-R-4.4.1
</span><span id="__span-25-2"><a id="__codelineno-25-2" name="__codelineno-25-2" href="#__codelineno-25-2"></a>
</span><span id="__span-25-3"><a id="__codelineno-25-3" name="__codelineno-25-3" href="#__codelineno-25-3"></a>---------------------------------------------------------------------------------
</span><span id="__span-25-4"><a id="__codelineno-25-4" name="__codelineno-25-4" href="#__codelineno-25-4"></a><span class="w">  </span>R-bundle-Bioconductor:<span class="w"> </span>R-bundle-Bioconductor/3.18-R-4.4.1
</span><span id="__span-25-5"><a id="__codelineno-25-5" name="__codelineno-25-5" href="#__codelineno-25-5"></a>---------------------------------------------------------------------------------
</span><span id="__span-25-6"><a id="__codelineno-25-6" name="__codelineno-25-6" href="#__codelineno-25-6"></a><span class="w">    </span>Description:
</span><span id="__span-25-7"><a id="__codelineno-25-7" name="__codelineno-25-7" href="#__codelineno-25-7"></a><span class="w">      </span>Bioconductor<span class="w"> </span>provides<span class="w"> </span>tools<span class="w"> </span><span class="k">for</span><span class="w"> </span>the<span class="w"> </span>analysis<span class="w"> </span>and<span class="w"> </span>coprehension<span class="w"> </span>of
</span><span id="__span-25-8"><a id="__codelineno-25-8" name="__codelineno-25-8" href="#__codelineno-25-8"></a><span class="w">      </span>high-throughput<span class="w"> </span>genomic<span class="w"> </span>data.
</span><span id="__span-25-9"><a id="__codelineno-25-9" name="__codelineno-25-9" href="#__codelineno-25-9"></a>
</span><span id="__span-25-10"><a id="__codelineno-25-10" name="__codelineno-25-10" href="#__codelineno-25-10"></a>
</span><span id="__span-25-11"><a id="__codelineno-25-11" name="__codelineno-25-11" href="#__codelineno-25-11"></a><span class="w">    </span>You<span class="w"> </span>will<span class="w"> </span>need<span class="w"> </span>to<span class="w"> </span>load<span class="w"> </span>all<span class="w"> </span>module<span class="o">(</span>s<span class="o">)</span><span class="w"> </span>on<span class="w"> </span>any<span class="w"> </span>one<span class="w"> </span>of<span class="w"> </span>the<span class="w"> </span>lines<span class="w"> </span>below<span class="w"> </span>before<span class="w"> </span>the<span class="w"> </span><span class="s2">&quot;R-b</span>
</span><span id="__span-25-12"><a id="__codelineno-25-12" name="__codelineno-25-12" href="#__codelineno-25-12"></a><span class="s2">undle-Bioconductor/3.18-R-4.4.1&quot;</span><span class="w"> </span>module<span class="w"> </span>is<span class="w"> </span>available<span class="w"> </span>to<span class="w"> </span>load.
</span><span id="__span-25-13"><a id="__codelineno-25-13" name="__codelineno-25-13" href="#__codelineno-25-13"></a>
</span><span id="__span-25-14"><a id="__codelineno-25-14" name="__codelineno-25-14" href="#__codelineno-25-14"></a><span class="w">      </span>GCC/12.3.0<span class="w">  </span>OpenMPI/4.1.5
</span><span id="__span-25-15"><a id="__codelineno-25-15" name="__codelineno-25-15" href="#__codelineno-25-15"></a>
</span><span id="__span-25-16"><a id="__codelineno-25-16" name="__codelineno-25-16" href="#__codelineno-25-16"></a><span class="w">    </span>Help:
</span><span id="__span-25-17"><a id="__codelineno-25-17" name="__codelineno-25-17" href="#__codelineno-25-17"></a>
</span><span id="__span-25-18"><a id="__codelineno-25-18" name="__codelineno-25-18" href="#__codelineno-25-18"></a><span class="w">      </span><span class="nv">Description</span>
</span><span id="__span-25-19"><a id="__codelineno-25-19" name="__codelineno-25-19" href="#__codelineno-25-19"></a><span class="w">      </span><span class="o">===========</span>
</span><span id="__span-25-20"><a id="__codelineno-25-20" name="__codelineno-25-20" href="#__codelineno-25-20"></a><span class="w">      </span>Bioconductor<span class="w"> </span>provides<span class="w"> </span>tools<span class="w"> </span><span class="k">for</span><span class="w"> </span>the<span class="w"> </span>analysis<span class="w"> </span>and<span class="w"> </span>coprehension
</span><span id="__span-25-21"><a id="__codelineno-25-21" name="__codelineno-25-21" href="#__codelineno-25-21"></a><span class="w">       </span>of<span class="w"> </span>high-throughput<span class="w"> </span>genomic<span class="w"> </span>data.
</span><span id="__span-25-22"><a id="__codelineno-25-22" name="__codelineno-25-22" href="#__codelineno-25-22"></a>
</span><span id="__span-25-23"><a id="__codelineno-25-23" name="__codelineno-25-23" href="#__codelineno-25-23"></a>
</span><span id="__span-25-24"><a id="__codelineno-25-24" name="__codelineno-25-24" href="#__codelineno-25-24"></a><span class="w">      </span>More<span class="w"> </span><span class="nv">information</span>
</span><span id="__span-25-25"><a id="__codelineno-25-25" name="__codelineno-25-25" href="#__codelineno-25-25"></a><span class="w">      </span><span class="o">================</span>
</span><span id="__span-25-26"><a id="__codelineno-25-26" name="__codelineno-25-26" href="#__codelineno-25-26"></a><span class="w">       </span>-<span class="w"> </span>Homepage:<span class="w"> </span>https://bioconductor.org
</span><span id="__span-25-27"><a id="__codelineno-25-27" name="__codelineno-25-27" href="#__codelineno-25-27"></a>
</span><span id="__span-25-28"><a id="__codelineno-25-28" name="__codelineno-25-28" href="#__codelineno-25-28"></a>
</span><span id="__span-25-29"><a id="__codelineno-25-29" name="__codelineno-25-29" href="#__codelineno-25-29"></a><span class="w">      </span>Included<span class="w"> </span><span class="nv">extensions</span>
</span><span id="__span-25-30"><a id="__codelineno-25-30" name="__codelineno-25-30" href="#__codelineno-25-30"></a><span class="w">      </span><span class="o">===================</span>
</span><span id="__span-25-31"><a id="__codelineno-25-31" name="__codelineno-25-31" href="#__codelineno-25-31"></a><span class="w">      </span>affxparser-1.74.0,<span class="w"> </span>affy-1.80.0,<span class="w"> </span>affycoretools-1.74.0,<span class="w"> </span>affyio-1.72.0,
</span><span id="__span-25-32"><a id="__codelineno-25-32" name="__codelineno-25-32" href="#__codelineno-25-32"></a><span class="w">      </span>AgiMicroRna-2.52.0,<span class="w"> </span>agricolae-1.3-7,<span class="w"> </span>ALDEx2-1.34.0,<span class="w"> </span>ALL-1.44.0,<span class="w"> </span>ANCOMBC-2.4.0,
</span><span id="__span-25-33"><a id="__codelineno-25-33" name="__codelineno-25-33" href="#__codelineno-25-33"></a><span class="w">      </span>annaffy-1.74.0,<span class="w"> </span>annotate-1.80.0,<span class="w"> </span>AnnotationDbi-1.64.1,
</span><span id="__span-25-34"><a id="__codelineno-25-34" name="__codelineno-25-34" href="#__codelineno-25-34"></a><span class="w">      </span>AnnotationFilter-1.26.0,<span class="w"> </span>AnnotationForge-1.44.0,<span class="w"> </span>AnnotationHub-3.10.0,
</span><span id="__span-25-35"><a id="__codelineno-25-35" name="__codelineno-25-35" href="#__codelineno-25-35"></a><span class="w">      </span>anytime-0.3.9,<span class="w"> </span>aroma.affymetrix-3.2.1,<span class="w"> </span>aroma.apd-0.7.0,<span class="w"> </span>aroma.core-3.3.0,
</span><span id="__span-25-36"><a id="__codelineno-25-36" name="__codelineno-25-36" href="#__codelineno-25-36"></a><span class="w">      </span>aroma.light-3.32.0,<span class="w"> </span>ash-1.0-15,<span class="w"> </span>ATACseqQC-1.26.0,<span class="w"> </span>AUCell-1.24.0,
</span><span id="__span-25-37"><a id="__codelineno-25-37" name="__codelineno-25-37" href="#__codelineno-25-37"></a><span class="w">      </span>aws.s3-0.3.21,<span class="w"> </span>aws.signature-0.6.0,<span class="w"> </span>babelgene-22.9,<span class="w"> </span>ballgown-2.34.0,
</span><span id="__span-25-38"><a id="__codelineno-25-38" name="__codelineno-25-38" href="#__codelineno-25-38"></a><span class="w">      </span>basilisk-1.14.2,<span class="w"> </span>basilisk.utils-1.14.1,<span class="w"> </span>batchelor-1.18.1,<span class="w"> </span>baySeq-2.36.0,
</span><span id="__span-25-39"><a id="__codelineno-25-39" name="__codelineno-25-39" href="#__codelineno-25-39"></a><span class="w">      </span>beachmat-2.18.0,<span class="w"> </span>BH-1.84.0-0,<span class="w"> </span>Biobase-2.62.0,<span class="w"> </span>BiocBaseUtils-1.4.0,<span class="w"> </span>...
</span></code></pre></div>
<p>The list of extensions is too long to copy here, but some popular extensions included in this module are: DeSeq2, GenomeInfoDb, MStats, Seurat, Rsamtools, and more.</p>
<p>The above prerequisites and the main package can be loaded either one at a time or all at once with,</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-26-1"><a id="__codelineno-26-1" name="__codelineno-26-1" href="#__codelineno-26-1"></a>$<span class="w"> </span>ml<span class="w"> </span>GCC/12.3.0<span class="w">  </span>OpenMPI/4.1.5<span class="w">  </span>R-bundle-Bioconductor/3.18-R-4.4.1
</span></code></pre></div>
<p>In this case, R-bundle-Bioconductor loads the version of R that it is based on automatically (along with about 130 other modules!). That is not the case for all R-bundles at all HPC centres, so pay attention to the prerequisites.</p>
</details>
</div>
<h3 id="matlab__and__another__tool__you__will__be__using">Matlab and another tool you will be using<a class="headerlink" href="#matlab__and__another__tool__you__will__be__using" title="Permanent link">&para;</a></h3>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 1: Try to find documentation of the program on your cluster</p>
<details class="note">
<summary>Documentation for MATLAB</summary>
<ul>
<li>Documentation at the HPC centres UPPMAX, HPC2N, LUNARC, NSC, and PDC</li>
<li><a href="http://docs.uppmax.uu.se/software/matlab/">UPPMAX Matlab docs</a></li>
<li>HPC2N: <a href="https://www.hpc2n.umu.se/resources/software/matlab">Matlab docs</a> and <a href="https://www.hpc2n.umu.se/documentation/guides/parallel-matlab">parallel Matlab docs</a></li>
<li><a href="https://lunarc-documentation.readthedocs.io/en/latest/guides/applications/MATLAB/">LUNARC Matlab docs</a></li>
<li>NSC: click <a href="https://www.nsc.liu.se/software/docs/matlab/">here for general instructions</a> and <a href="https://www.nsc.liu.se/software/catalogue/tetralith/modules/matlab.html">here for installations on Tetralith</a> specifically</li>
<li><a href="https://support.pdc.kth.se/doc/applications/?sub=matlab/">PDC Matlab docs</a></li>
</ul>
</details>
</div>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 2: Find version of it. Is it installed?? Look for other tools until you find one that is installed</p>
</div>
<details class="note">
<summary>Output from <code>ml avail matlab</code></summary>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-27-1"><a id="__codelineno-27-1" name="__codelineno-27-1" href="#__codelineno-27-1"></a>$<span class="w"> </span>ml<span class="w"> </span>avail<span class="w"> </span>matlab
</span><span id="__span-27-2"><a id="__codelineno-27-2" name="__codelineno-27-2" href="#__codelineno-27-2"></a>
</span><span id="__span-27-3"><a id="__codelineno-27-3" name="__codelineno-27-3" href="#__codelineno-27-3"></a>---------------------<span class="w"> </span>/software/sse2/tetralith_el9/modules<span class="w"> </span>---------------------
</span><span id="__span-27-4"><a id="__codelineno-27-4" name="__codelineno-27-4" href="#__codelineno-27-4"></a><span class="w">   </span>MATLAB/recommendation<span class="w"> </span><span class="o">(</span>D<span class="o">)</span><span class="w">    </span>MATLAB/2024a-hpc1-bdist
</span><span id="__span-27-5"><a id="__codelineno-27-5" name="__codelineno-27-5" href="#__codelineno-27-5"></a><span class="w">   </span>MATLAB/2023a-bdist<span class="w">           </span>MATLAB/2025a-hpc1-bdist
</span><span id="__span-27-6"><a id="__codelineno-27-6" name="__codelineno-27-6" href="#__codelineno-27-6"></a><span class="w">   </span>MATLAB/2023b-bdist
</span><span id="__span-27-7"><a id="__codelineno-27-7" name="__codelineno-27-7" href="#__codelineno-27-7"></a>
</span><span id="__span-27-8"><a id="__codelineno-27-8" name="__codelineno-27-8" href="#__codelineno-27-8"></a><span class="w">  </span>Where:
</span><span id="__span-27-9"><a id="__codelineno-27-9" name="__codelineno-27-9" href="#__codelineno-27-9"></a><span class="w">   </span>D:<span class="w">  </span>Default<span class="w"> </span>Module
</span></code></pre></div>
<p>Once you have chosen a specific version, use <code>ml spider</code> to check if there are prerequisites, like so:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-28-1"><a id="__codelineno-28-1" name="__codelineno-28-1" href="#__codelineno-28-1"></a>$<span class="w"> </span>ml<span class="w"> </span>spider<span class="w"> </span>MATLAB/2024a-hpc1-bdist
</span></code></pre></div>
<p>The full output is too verbose to reprint in full here, but the one important line reads: </p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-29-1"><a id="__codelineno-29-1" name="__codelineno-29-1" href="#__codelineno-29-1"></a><span class="w">    </span>This<span class="w"> </span>module<span class="w"> </span>can<span class="w"> </span>be<span class="w"> </span>loaded<span class="w"> </span>directly:<span class="w"> </span>module<span class="w"> </span>load<span class="w"> </span>MATLAB/2024a-hpc1-bdist
</span></code></pre></div>
<p>The command after the colon (:) can be copied, pasted, and entered directly into the bash prompt to load the module, or you can type the short version as follows:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-30-1"><a id="__codelineno-30-1" name="__codelineno-30-1" href="#__codelineno-30-1"></a>$<span class="w"> </span>ml<span class="w"> </span>MATLAB/2024a-hpc1-bdist
</span></code></pre></div>
</details>
<details class="note">
<summary>Example OpenFOAM at Cosmos</summary>
<p>As usual, we start by checking the versions available with, eg. <code>ml spider OpenFOAM</code> (or <code>ml avail OpenFOAM</code> at a facility like NSC where all modules are visible regardless of the presence of prerequisites). </p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-31-1"><a id="__codelineno-31-1" name="__codelineno-31-1" href="#__codelineno-31-1"></a>$<span class="w"> </span>ml<span class="w"> </span>spider<span class="w"> </span>OpenFOAM
</span><span id="__span-31-2"><a id="__codelineno-31-2" name="__codelineno-31-2" href="#__codelineno-31-2"></a>
</span><span id="__span-31-3"><a id="__codelineno-31-3" name="__codelineno-31-3" href="#__codelineno-31-3"></a>--------------------------------------------------------------------------------------
</span><span id="__span-31-4"><a id="__codelineno-31-4" name="__codelineno-31-4" href="#__codelineno-31-4"></a><span class="w">  </span>OpenFOAM:
</span><span id="__span-31-5"><a id="__codelineno-31-5" name="__codelineno-31-5" href="#__codelineno-31-5"></a>--------------------------------------------------------------------------------------
</span><span id="__span-31-6"><a id="__codelineno-31-6" name="__codelineno-31-6" href="#__codelineno-31-6"></a><span class="w">    </span>Description:
</span><span id="__span-31-7"><a id="__codelineno-31-7" name="__codelineno-31-7" href="#__codelineno-31-7"></a><span class="w">      </span>OpenFOAM<span class="w"> </span>is<span class="w"> </span>a<span class="w"> </span>free,<span class="w"> </span>open<span class="w"> </span><span class="nb">source</span><span class="w"> </span>CFD<span class="w"> </span>software<span class="w"> </span>package.<span class="w"> </span>OpenFOAM<span class="w"> </span>has<span class="w"> </span>an<span class="w"> </span>extensive
</span><span id="__span-31-8"><a id="__codelineno-31-8" name="__codelineno-31-8" href="#__codelineno-31-8"></a><span class="w">      </span>range<span class="w"> </span>of<span class="w"> </span>features<span class="w"> </span>to<span class="w"> </span>solve<span class="w"> </span>anything<span class="w"> </span>from<span class="w"> </span>complex<span class="w"> </span>fluid<span class="w"> </span>flows<span class="w"> </span>involving<span class="w"> </span>chemical
</span><span id="__span-31-9"><a id="__codelineno-31-9" name="__codelineno-31-9" href="#__codelineno-31-9"></a><span class="w">      </span>reactions,<span class="w"> </span>turbulence<span class="w"> </span>and<span class="w"> </span>heat<span class="w"> </span>transfer,<span class="w"> </span>to<span class="w"> </span>solid<span class="w"> </span>dynamics<span class="w"> </span>and
</span><span id="__span-31-10"><a id="__codelineno-31-10" name="__codelineno-31-10" href="#__codelineno-31-10"></a><span class="w">      </span>electromagnetics.
</span><span id="__span-31-11"><a id="__codelineno-31-11" name="__codelineno-31-11" href="#__codelineno-31-11"></a>
</span><span id="__span-31-12"><a id="__codelineno-31-12" name="__codelineno-31-12" href="#__codelineno-31-12"></a><span class="w">     </span>Versions:
</span><span id="__span-31-13"><a id="__codelineno-31-13" name="__codelineno-31-13" href="#__codelineno-31-13"></a><span class="w">        </span>OpenFOAM/v2112
</span><span id="__span-31-14"><a id="__codelineno-31-14" name="__codelineno-31-14" href="#__codelineno-31-14"></a><span class="w">        </span>OpenFOAM/v2206
</span><span id="__span-31-15"><a id="__codelineno-31-15" name="__codelineno-31-15" href="#__codelineno-31-15"></a><span class="w">        </span>OpenFOAM/v2306
</span><span id="__span-31-16"><a id="__codelineno-31-16" name="__codelineno-31-16" href="#__codelineno-31-16"></a><span class="w">        </span>OpenFOAM/v2406
</span><span id="__span-31-17"><a id="__codelineno-31-17" name="__codelineno-31-17" href="#__codelineno-31-17"></a><span class="w">        </span>OpenFOAM/7-20200508
</span><span id="__span-31-18"><a id="__codelineno-31-18" name="__codelineno-31-18" href="#__codelineno-31-18"></a><span class="w">        </span>OpenFOAM/9
</span><span id="__span-31-19"><a id="__codelineno-31-19" name="__codelineno-31-19" href="#__codelineno-31-19"></a><span class="w">        </span>OpenFOAM/10
</span><span id="__span-31-20"><a id="__codelineno-31-20" name="__codelineno-31-20" href="#__codelineno-31-20"></a><span class="w">        </span>OpenFOAM/11
</span><span id="__span-31-21"><a id="__codelineno-31-21" name="__codelineno-31-21" href="#__codelineno-31-21"></a>
</span><span id="__span-31-22"><a id="__codelineno-31-22" name="__codelineno-31-22" href="#__codelineno-31-22"></a>--------------------------------------------------------------------------------------
</span><span id="__span-31-23"><a id="__codelineno-31-23" name="__codelineno-31-23" href="#__codelineno-31-23"></a><span class="w">  </span>For<span class="w"> </span>detailed<span class="w"> </span>information<span class="w"> </span>about<span class="w"> </span>a<span class="w"> </span>specific<span class="w"> </span><span class="s2">&quot;OpenFOAM&quot;</span><span class="w"> </span>package<span class="w"> </span><span class="o">(</span>including<span class="w"> </span>how<span class="w"> </span>to<span class="w"> </span>load<span class="w"> </span>the<span class="w"> </span>
</span><span id="__span-31-24"><a id="__codelineno-31-24" name="__codelineno-31-24" href="#__codelineno-31-24"></a>modules<span class="o">)</span><span class="w"> </span>use<span class="w"> </span>the<span class="w"> </span>module<span class="err">&#39;</span>s<span class="w"> </span>full<span class="w"> </span>name.
</span><span id="__span-31-25"><a id="__codelineno-31-25" name="__codelineno-31-25" href="#__codelineno-31-25"></a><span class="w">  </span>Note<span class="w"> </span>that<span class="w"> </span>names<span class="w"> </span>that<span class="w"> </span>have<span class="w"> </span>a<span class="w"> </span>trailing<span class="w"> </span><span class="o">(</span>E<span class="o">)</span><span class="w"> </span>are<span class="w"> </span>extensions<span class="w"> </span>provided<span class="w"> </span>by<span class="w"> </span>other<span class="w"> </span>modules.
</span><span id="__span-31-26"><a id="__codelineno-31-26" name="__codelineno-31-26" href="#__codelineno-31-26"></a><span class="w">  </span>For<span class="w"> </span>example:
</span><span id="__span-31-27"><a id="__codelineno-31-27" name="__codelineno-31-27" href="#__codelineno-31-27"></a>
</span><span id="__span-31-28"><a id="__codelineno-31-28" name="__codelineno-31-28" href="#__codelineno-31-28"></a><span class="w">     </span>$<span class="w"> </span>module<span class="w"> </span>spider<span class="w"> </span>OpenFOAM/11
</span><span id="__span-31-29"><a id="__codelineno-31-29" name="__codelineno-31-29" href="#__codelineno-31-29"></a>--------------------------------------------------------------------------------------
</span></code></pre></div>
<p>Let us look at a recent version:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-32-1"><a id="__codelineno-32-1" name="__codelineno-32-1" href="#__codelineno-32-1"></a>$<span class="w"> </span>ml<span class="w"> </span>spider<span class="w"> </span>OpenFOAM/11
</span><span id="__span-32-2"><a id="__codelineno-32-2" name="__codelineno-32-2" href="#__codelineno-32-2"></a>--------------------------------------------------------------------------------------
</span><span id="__span-32-3"><a id="__codelineno-32-3" name="__codelineno-32-3" href="#__codelineno-32-3"></a><span class="w">  </span>OpenFOAM:<span class="w"> </span>OpenFOAM/11
</span><span id="__span-32-4"><a id="__codelineno-32-4" name="__codelineno-32-4" href="#__codelineno-32-4"></a>--------------------------------------------------------------------------------------
</span><span id="__span-32-5"><a id="__codelineno-32-5" name="__codelineno-32-5" href="#__codelineno-32-5"></a><span class="w">    </span>Description:
</span><span id="__span-32-6"><a id="__codelineno-32-6" name="__codelineno-32-6" href="#__codelineno-32-6"></a><span class="w">      </span>OpenFOAM<span class="w"> </span>is<span class="w"> </span>a<span class="w"> </span>free,<span class="w"> </span>open<span class="w"> </span><span class="nb">source</span><span class="w"> </span>CFD<span class="w"> </span>software<span class="w"> </span>package.<span class="w"> </span>OpenFOAM<span class="w"> </span>has<span class="w"> </span>an<span class="w"> </span>extensive
</span><span id="__span-32-7"><a id="__codelineno-32-7" name="__codelineno-32-7" href="#__codelineno-32-7"></a><span class="w">      </span>range<span class="w"> </span>of<span class="w"> </span>features<span class="w"> </span>to<span class="w"> </span>solve<span class="w"> </span>anything<span class="w"> </span>from<span class="w"> </span>complex<span class="w"> </span>fluid<span class="w"> </span>flows<span class="w"> </span>involving<span class="w"> </span>chemical
</span><span id="__span-32-8"><a id="__codelineno-32-8" name="__codelineno-32-8" href="#__codelineno-32-8"></a><span class="w">      </span>reactions,<span class="w"> </span>turbulence<span class="w"> </span>and<span class="w"> </span>heat<span class="w"> </span>transfer,<span class="w"> </span>to<span class="w"> </span>solid<span class="w"> </span>dynamics<span class="w"> </span>and
</span><span id="__span-32-9"><a id="__codelineno-32-9" name="__codelineno-32-9" href="#__codelineno-32-9"></a><span class="w">      </span>electromagnetics.
</span><span id="__span-32-10"><a id="__codelineno-32-10" name="__codelineno-32-10" href="#__codelineno-32-10"></a>
</span><span id="__span-32-11"><a id="__codelineno-32-11" name="__codelineno-32-11" href="#__codelineno-32-11"></a>
</span><span id="__span-32-12"><a id="__codelineno-32-12" name="__codelineno-32-12" href="#__codelineno-32-12"></a><span class="w">    </span>You<span class="w"> </span>will<span class="w"> </span>need<span class="w"> </span>to<span class="w"> </span>load<span class="w"> </span>all<span class="w"> </span>module<span class="o">(</span>s<span class="o">)</span><span class="w"> </span>on<span class="w"> </span>any<span class="w"> </span>one<span class="w"> </span>of<span class="w"> </span>the<span class="w"> </span>lines<span class="w"> </span>below<span class="w"> </span>before<span class="w"> </span>the<span class="w"> </span><span class="s2">&quot;OpenFOAM</span>
</span><span id="__span-32-13"><a id="__codelineno-32-13" name="__codelineno-32-13" href="#__codelineno-32-13"></a><span class="s2">/11&quot;</span><span class="w"> </span>module<span class="w"> </span>is<span class="w"> </span>available<span class="w"> </span>to<span class="w"> </span>load.
</span><span id="__span-32-14"><a id="__codelineno-32-14" name="__codelineno-32-14" href="#__codelineno-32-14"></a>
</span><span id="__span-32-15"><a id="__codelineno-32-15" name="__codelineno-32-15" href="#__codelineno-32-15"></a><span class="w">      </span>GCC/11.3.0<span class="w">  </span>OpenMPI/4.1.4
</span><span id="__span-32-16"><a id="__codelineno-32-16" name="__codelineno-32-16" href="#__codelineno-32-16"></a>
</span><span id="__span-32-17"><a id="__codelineno-32-17" name="__codelineno-32-17" href="#__codelineno-32-17"></a><span class="w">    </span>Help:
</span><span id="__span-32-18"><a id="__codelineno-32-18" name="__codelineno-32-18" href="#__codelineno-32-18"></a><span class="w">      </span><span class="nv">Description</span>
</span><span id="__span-32-19"><a id="__codelineno-32-19" name="__codelineno-32-19" href="#__codelineno-32-19"></a><span class="w">      </span><span class="o">===========</span>
</span><span id="__span-32-20"><a id="__codelineno-32-20" name="__codelineno-32-20" href="#__codelineno-32-20"></a><span class="w">      </span>OpenFOAM<span class="w"> </span>is<span class="w"> </span>a<span class="w"> </span>free,<span class="w"> </span>open<span class="w"> </span><span class="nb">source</span><span class="w"> </span>CFD<span class="w"> </span>software<span class="w"> </span>package.
</span><span id="__span-32-21"><a id="__codelineno-32-21" name="__codelineno-32-21" href="#__codelineno-32-21"></a><span class="w">       </span>OpenFOAM<span class="w"> </span>has<span class="w"> </span>an<span class="w"> </span>extensive<span class="w"> </span>range<span class="w"> </span>of<span class="w"> </span>features<span class="w"> </span>to<span class="w"> </span>solve
</span><span id="__span-32-22"><a id="__codelineno-32-22" name="__codelineno-32-22" href="#__codelineno-32-22"></a><span class="w">       </span>anything<span class="w"> </span>from<span class="w"> </span>complex<span class="w"> </span>fluid<span class="w"> </span>flows<span class="w"> </span>involving<span class="w"> </span>chemical
</span><span id="__span-32-23"><a id="__codelineno-32-23" name="__codelineno-32-23" href="#__codelineno-32-23"></a><span class="w">       </span>reactions,<span class="w"> </span>turbulence<span class="w"> </span>and<span class="w"> </span>heat<span class="w"> </span>transfer,<span class="w"> </span>to<span class="w"> </span>solid<span class="w"> </span>dynamics
</span><span id="__span-32-24"><a id="__codelineno-32-24" name="__codelineno-32-24" href="#__codelineno-32-24"></a><span class="w">       </span>and<span class="w"> </span>electromagnetics.
</span><span id="__span-32-25"><a id="__codelineno-32-25" name="__codelineno-32-25" href="#__codelineno-32-25"></a>
</span><span id="__span-32-26"><a id="__codelineno-32-26" name="__codelineno-32-26" href="#__codelineno-32-26"></a>
</span><span id="__span-32-27"><a id="__codelineno-32-27" name="__codelineno-32-27" href="#__codelineno-32-27"></a><span class="w">      </span>More<span class="w"> </span><span class="nv">information</span>
</span><span id="__span-32-28"><a id="__codelineno-32-28" name="__codelineno-32-28" href="#__codelineno-32-28"></a><span class="w">      </span><span class="o">================</span>
</span><span id="__span-32-29"><a id="__codelineno-32-29" name="__codelineno-32-29" href="#__codelineno-32-29"></a><span class="w">       </span>-<span class="w"> </span>Homepage:<span class="w"> </span>https://www.openfoam.org/
</span></code></pre></div>
<p>On this system, GCC and OpenMPI must be loaded first, but this is not true for every system. Indeed, some versions on some systems (e.g. at NSC) load and use the compilers, MPI libraries, and mathematics libraries from Intel toolchains.</p>
<p>Now we can load everything all at once like so:</p>
<div class="language-bash highlight"><pre><span></span><code><span id="__span-33-1"><a id="__codelineno-33-1" name="__codelineno-33-1" href="#__codelineno-33-1"></a>$<span class="w"> </span>ml<span class="w"> </span>GCC/11.3.0<span class="w">  </span>OpenMPI/4.1.4<span class="w"> </span>OpenFOAM/11
</span></code></pre></div>
<p>or load each one at a time. The above command loads almost 90 modules, including several Python packages and visualisation libraries, all of which can be viewed by entering <code>ml</code>.</p>
</details>
<div class="admonition note">
<p class="admonition-title"> <img src="../images/shell-logo_small.png"> Exercise 3. Load and start it</p>
<p>At most HPC centres, Matlab and other tools can be loaded directly, but PDC requires the usual prerequisite <code>PDC/XX.XX</code> or <code>PDCOLD/XX.XX</code>. Capitalisation and other naming conventions also vary between HPC centres; for more information, refer to <a href="https://uppmax.github.io/R-matlab-julia-HPC/matlab/load_runMatlab/#check-for-matlab-versions">this section of the R, Matlab, and Julia for HPC course</a>.</p>
</div>
<h2 id="wrap-up__with__questions__no__recording">Wrap-up with questions (No recording)<a class="headerlink" href="#wrap-up__with__questions__no__recording" title="Permanent link">&para;</a></h2>
<ul>
<li>Did it work out?</li>
<li>Would you like a tool installed?</li>
<li>Questions?</li>
</ul>
<h2 id="courses">Courses<a class="headerlink" href="#courses" title="Permanent link">&para;</a></h2>
<ul>
<li><a href="https://uppmax.github.io/HPC-python/index.html">Using Python in an HPC environment</a></li>
<li><a href="https://uppmax.github.io/R-matlab-julia-HPC/">R-MATLAB-Julia</a></li>
</ul>
<div class="admonition abstract">
<p class="admonition-title">Summary</p>
<ul>
<li>Many Python and R packages come in bundle modules.<ul>
<li>Python has &gt;10 of them</li>
<li>R has basically 2: CRAN and Bioconductor.</li>
<li>Load a bundle and the correct version of Python or R is loaded on the fly.</li>
</ul>
</li>
<li>All clusters have RStudio and Jupyter installed in some way. <ul>
<li>Some clusters have Spyder and VS Code</li>
</ul>
</li>
<li>Some tools are licensed.</li>
</ul>
</div>
              
            </div>
          </div><footer>
    <div class="rst-footer-buttons" role="navigation" aria-label="Footer Navigation">
        <a href="../toolchains/" class="btn btn-neutral float-left" title="Compiler toolchains"><span class="icon icon-circle-arrow-left"></span> Previous</a>
        <a href="../batch/" class="btn btn-neutral float-right" title="Modules in batch scripts">Next <span class="icon icon-circle-arrow-right"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
  </div>

  Built with <a href="https://www.mkdocs.org/">MkDocs</a> using a <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
          
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" aria-label="Versions">
  <span class="rst-current-version" data-toggle="rst-current-version">
    
    
      <span><a href="../toolchains/" style="color: #fcfcfc">&laquo; Previous</a></span>
    
    
      <span><a href="../batch/" style="color: #fcfcfc">Next &raquo;</a></span>
    
  </span>
</div>
    <script src="../js/jquery-3.6.0.min.js"></script>
    <script>var base_url = "..";</script>
    <script src="../js/theme_extra.js"></script>
    <script src="../js/theme.js"></script>
      <script src="../assets/_markdown_exec_pyodide.js"></script>
      <script src="../search/main.js"></script>
      <script src="../js/open_in_new_tab.js"></script>
    <script>
        jQuery(function () {
            SphinxRtdTheme.Navigation.enable(true);
        });
    </script>

</body>
</html>
